Add format specification in the docs (#14)
sebapersson/petab_sciml: docs/src/format.md

A PEtab SciML problem can also include multiple neural networks. Aside from ensuring that neural networks do not conflict (e.g., by sharing the same output), no special considerations are required. Each additional network is included just as it would be in the single-network case.

## [Neural Network Model Format](@id net_format)

The neural network model format is flexible, meaning models can be provided in any format compatible with the PEtab SciML specification (for example, [Lux.jl](https://github.com/LuxDL/Lux.jl) in [PEtab.jl](https://github.com/sebapersson/PEtab.jl)). Additionally, the `petab_sciml` library provides a neural network YAML file format that can be imported by tools across various programming languages. This format flexibility exists because, although the YAML format can accommodate many architectures, some may still be difficult to represent. However, the YAML format is recommended whenever possible to facilitate model exchange.

A neural network model must consist of two parts to be compatible with the PEtab SciML specification:

> - **layers**: A constructor that defines the network layers, each with a unique identifier.

Author: You mean just have something like: Defines the network layers, each with a unique identifier.

- **forward**: A forward pass function that, given input arguments, specifies the order in which layers are called, applies any activation functions, and returns one or several arrays. The forward function can accept more than one input argument (`n > 1`), and in the [mapping table](@ref mapping_table), the forward function's `n`th input argument (ignoring any potential class arguments such as `self`) is referred to as `inputArgumentIndex{n}`. Similar holds for the output. Aside from the neural network output values, every component that should be visible to other parts of the PEtab SciML problem must be defined elsewhere (e.g., in **layers**).

### [Neural Network Parameters](@id hdf5_ps_structure)

All parameters for a neural network model are stored in an HDF5 file, with the file path specified in the problem [YAML file](@ref YAML_file). The HDF5 parameter file is expected to have the following structure for an arbitrary number of layers:

```hdf5
parameters.hdf5

> └───layerId{1} (group)

Author: Should be `layerId1` (my mistake)

│   ├── arrayId{1}
│   └── arrayId{2}
└───layerId{2} (group)
    ├── arrayId{1}
    └── arrayId{2}
```

> The indexing convention and naming for `arrayId` depend on the neural network model library:

Author: This is an interesting point. Starting with the naming, there is not way I think we can avoid this. 

When the model is in the YAML format things are easy, because then per the standard we enforce the PyTorch array names. 

However, lets say that it is now a Julia user providing the neural network in Lux.jl, and they have a layer not covered by the YAML standard, and on parameter array is named `scale` (this does happen). Here enforcing the naming to be `weight` would be confusing for the Julia user (and would it even make sense). Similar for the indexing, forcing a Julia user providing the network in Julia as Lux.jl to use PyTorch indexing would be confusing (as array ordering really is different, I need mapping functions for this in PEtab.jl for YAML supported layers). Also generating the HDF5 file is not a problem, as in neural network libraries one can access the array ids.

In short, as long as we are flexible in neural network format, which means we allow packages with different naming and indexing conventions for array-ids, I argue this is the least confusing solution (as the user gets what they expect from the neural network package they use, and required files can be generated). We can add that packages should provide helpers for generating proper files?

> The indexing convention and naming for `arrayId` depend on the neural network model library:

Reviewer: For interoperability, we could:
- include the array order (e.g. row-major / column-major) as some metadata inside the HDF5
- have an opinionated way to flatten arrays such that both the weights and biases of a Linear module can be accessed via `net1.layers[linear1].parameters[i]`. We could say that when using `.parameters`, then all arrays are ordered alphabetically then flattened row-major. e.g. a Linear module with 5 inputs and 3 outputs with a 1x5 bias vector and 3x5 weight matrix could become a 1x20 parameter vector, where the first 5 value are the biases.
- if a user wants to use language-specific array orders, then they shouldn't use `.parameters` but rather `.weight` and `.bias`

I think this solves both problems: (1) users can write and use their problem in Julia, but then also (2) share it in a usable way since there is now sufficient information in the format itself to automatically convert the problem to a pytorch module.

This `.parameters[i]` indexing can be limited to the HDF5 format only, if you still want to avoid such indexing in the hybridization table etc (fine for me, I don't see an issue currently, unless users really want to estimate a single weight scalar inside a weight matrix...).

> The indexing convention and naming for `arrayId` depend on the neural network model library:

Author: > I think this solves both problems: (1) users can write and use their problem in Julia, but then also (2) share it in a usable way since there is now sufficient information in the format itself to automatically convert the problem to a pytorch module.

I might be missing something here, but if the user provides the neural network in another format (e.g. Lux.jl) do we not forgo guaranteed cross-language interoperability? Because, we still have that array ids are named differently (which will cause alphabetical mapping issues for exchange), and in Julia  some layers in Lux.jl (e.g. [scale](https://lux.csail.mit.edu/stable/api/Lux/layers)) do not have a corresponding variant in PyTorch.

Still, exchange of parameters would be nice, but in order to trust such exchange we would need tests for each layer type . These tests are available for the layers supported by the YAML standard. So how about we say that tools supporting a custom format should, for networks with all layers available in the YAML standard format, be able to export the parameters file in YAML format compatible format? This is also something that is straightforward to test for. 

Lastly, I do not want to have to complicated parameter structure, as I think for user friendliness the format should resemble the neural network model structure as close as possible (the HDF5 file should make sense if the user reads it directly)

> The indexing convention and naming for `arrayId` depend on the neural network model library:

Reviewer: > I agree that this should follow the specific nomenclature of the respective library (and by extension this also applies to the indexing).

Agreed. Is there other metadata than `order=row-major/column-major` that we would need to add to the HDF5 for this, to be able to use a Julia-generated HDF5 in Python?

> we still have that array ids are named differently

Ah, I guessed this would necessarily be resolved, because you would need to import the YAML format into a Lux model anyway, hence some mapping of layer names between libraries must exist? Or not, because the YAML format is completely optional?

> in Julia some layers in Lux.jl (e.g. [scale](https://lux.csail.mit.edu/stable/api/Lux/layers)) do not have a corresponding variant in PyTorch

Agreed, interoperability only makes sense for layers that we support in the MLModel YAML format. No need for full interoperability IMO

> So how about we say that tools supporting a custom format should, for networks with all layers available in the YAML standard format, be able to export the parameters file in YAML format compatible format?

:+1: Agreed, this is what I am suggesting too, to ensure that there is only one way to export the parameters. But your suggestion is nicer. For layers that are supported by the YAML format, then we should give an opinion of array names I guess?

> I do not want to have to complicated parameter structure

Also fine for me, but then is there a need to have `.parameters`? Some layers might name their parameters simply `parameters`, but is there a need for a generic `layer[layerId].parameters` on all layers? If so, how would indexing there work?

> The indexing convention and naming for `arrayId` depend on the neural network model library:

Author: > For layers that are supported by the YAML format, then we should give an opinion of array names I guess?

We already do. Some of the layers in Lux.jl have `scale` instead of `weight` as in PyTorch, and in the YAML format we always use PyTorch naming (and via testing I ensure everything works as expected). 

It is believed that the best overall approach for hiding the complexity of array naming is to have the HDF5 file follow the convention of the nerula network library. In addition, we should add in the specification that it is strongly encouraged for packages to include a function such as `export_ps_petab_sciml`, which, for exchange purposes, attempts to export the file to the YAML-compatible format and raises an error if any YAML-supported layers are present. I do not want to mix approaches for how to save the parameters  dependent on whether a specific layer is present in the YAML format, as this would for example be confusing if the user manually reads the file.

> Also fine for me, but then is there a need to have .parameters? Some layers might name their parameters simply parameters, but is there a need for a generic layer[layerId].parameters on all layers? If so, how would indexing there work?

If you mean `netId.parameters`, I think this is to match `netId.input ` and `netId.output` (consistent syntax). And with the current spec `layer[layerId].parameters `is not allowed. The way to index a layer array is: `netId.parameters[layerId].arrayId`, so having `arrayId = parameters` albeit weird, will work.

> The indexing convention and naming for `arrayId` depend on the neural network model library:

Reviewer: > have the HDF5 file follow the convention of the nerula network library

What is the benefit of the HDF5 parameters file in this case then? I can only see the benefit if it enables interoperability. If the HDF5 parameters file is only usable by one library, then it's just adding some burden to a library that might already support one of the other formats for inference time?

> The way to index a layer array is [...]

Alright thanks, I was confused from our previous conversation where I thought the syntax would be different. But fine for me either way

> The indexing convention and naming for `arrayId` depend on the neural network model library:

Author: > What is the benefit of the HDF5 parameters file in this case then?

The overall problem even though the neural network is custom, is still a PEtab problem (with the other files), and having a standardized way of storing parameters is the PEtab way and makes it easier to develop downstream pipelines etc

> The indexing convention and naming for `arrayId` depend on the neural network model library:

Reviewer: Just weird to me since it feels more like two standards (one for AMICI+Jax and one for PEtab.jl). I think the solution is to ensure there is sufficient metadata inside the HDF5 files such that, if another tool/framework has the same layer types, it can reconstruct the layers with their parameters, even if these layer types aren't in our NN YAML format.

This would mean that these arrays for unofficial layer types are stored with some `layerType` metadata like `{framework}:{layer type in framework}:{array type in framework}`, e.g. `layerType="Lux.jl:Linear:weight"`. This could also be a way to incorporate framework-specific layer types into the NN YAML format...

Anyway, we can discuss this again in the future HDF5 PR. I think there's no need to finalize all details here before that

- Neural network models in other formats follow the indexing and naming conventions of the respective package and programming language.

!!! tip "For developers: Respect memory order"
    Tools supporting the SciML extension should, for computational efficiency, reorder input data and potential layer parameter arrays to match the memory ordering of the target language. For example, PEtab.jl converts input data to follow Julia based indexing.

TODO: We will fix condition specific input in the YAML file later.

### [YAML Network file format](@id YAML_net_format)

The `petab_sciml` library provides a YAML neural network file format for model exchange. The YAML format follows PyTorch conventions for layer names and arguments. While YAML network files can be written manually, it is recommended approach to define a PyTorch `nn.Module` and use the `petab_sciml` library to automatically generate the YAML representation (see tutorials).

TODO: Should we have some description of the YAML format here? (can and should probably be added later)

## [Mapping Table](@id mapping_table)

> All neural networks are assigned an Id in the PEtab problem [YAML](@ref YAML_file) file. A neural network Id is not considered a valid PEtab identifier, to avoid confusion about what it refers to (e.g., parameters, inputs, outputs). Consequently, every neural network input, parameter, and output referenced in the PEtab problem must be defined under `modelEntityId` and mapped to a PEtab identifier. For the `PEtabEntityId` column the same rules as in PEtab v2 apply, and additionally array file Ids defined in the [YAML](@ref YAML_file) file are considered valid PEtab entities.

Reviewer 2: ```suggestion
All neural networks are assigned an identifier in the PEtab problem [YAML](@ref YAML_file) file. A neural network identifier is not considered a valid PEtab identifier, to avoid confusion about what it refers to (e.g., parameters, inputs, outputs). Consequently, every neural network input, parameter, and output referenced in the PEtab problem must be defined under `modelEntityId` and mapped to a PEtab identifier. For the `PEtabEntityId` column the same rules as in PEtab v2 apply, and additionally array file Ids defined in the [YAML](@ref YAML_file) file are considered valid PEtab entities.

```

### `modelEntityId` [STRING, REQUIRED]

A modeling-language-independent syntax which refers to inputs, outputs, and parameters of neural networks.

#### Parameters

The model Id `$netId.parameters[$layerId].{[$arrayId]{[$parameterIndex]}}` refers to the parameters of a neural network identified by `$netId`.

- `$layerId`: The unique identifier of the layer (e.g., `conv1`).
- `$arrayId`: The parameter array name specific to that layer (e.g., `weight`).
- `$parameterIndex`: The indexing into the parameter array ([syntax](@ref mapping_table_indexing)).

#### Inputs

The model Id `$netId.inputs{[$inputArgumentIndex]{[$inputIndex]}}` refers to specific inputs of the network identified by `$netId`.

> - `$inputArgumentIndex`: The input argument number in the neural network forward function. Starts from 1.

Reviewer 2: would prefer to not mix 1-based (here) and 0-based indexing (below)

| NON\_ESTIMATED\_ENTITY\_ID | MATH\_EXPRESSION |
| net1\_input1             | p1              |
| net1\_input2             | p1              |
| ...                     | ...             |

### Detailed Field Description

- `targetId` [NON\_ESTIMATED\_ENTITY\_ID, REQUIRED]: The identifier of the non-estimated entity that will be modified. Restrictions depend on hybridization type ([pre- or intra-simulation hybridization](@ref hybrid_types)). See below.
- `targetValue` [STRING, REQUIRED]: The value or expression that will be used to change the target.

### Pre-simulation hybridization

Pre-simulation neural network model inputs and outputs are constant targets (see case 1 [here](@ref hybrid_types)).

#### Inputs

Valid `targetValue`'s for a neural network input are:

- A parameter in the parameter table.
- An array input file (assigned an Id in the [YAML problem file](@ref YAML_file)).

#### Outputs

Valid `targetId`'s for a neural network output are:

- A non-estimated model parameter.
- A specie's initial value (referenced by the specie's Id). In this case, any other specie initialization is overridden.

#### Condition and Hybridization Tables

> Pre-simulation assignments can alternatively be made in the conditions table. Combinations are also permitted, for example, all inputs can be assigned in the condition table while all outputs are assigning in the hybridization table. **Importantly**, since the hybridization table defines assignments for all simulation conditions, any `targetId` assigned in the condition table cannot be assigned in the hybridization table, and vice versa.

Reviewer 2: I don't get what this is supposed to mean. I thought we don't want to allow assignments in the condition table. I think we should keep the structure fixed globally (mapping of inputs/outputs to PEtab IDs) but allow condition specific mapping of the _values_ of input variables.

Reviewer: I think this is what is meant, that the HDF5 input files are mapped/assigned to NN inputs in the condition table.

Or is the intended way to specify this condition-specific input data via the metadata in the HDF5 input data file? If so then I agree, inputs could be removed from this paragraph.

Author: This is badly worded from my side.

What I want to convey is that in the condition table we are allowed to assign values to the input variables, to make clear value assigning is not exclusive to the hybridization table  (as the input variables are categorized as NON_PARAMETER_TABLE_ID).

I will clarify this a bit.

Author: Taken together, this is what I want to say:

Neural network input variables are valid `targetId`s for the condition table as long as, following the PEtab standard, they are NON_PARAMETER_TABLE_ID. **Importantly**, since the hybridization table defines assignments for all simulation conditions, any `targetId` in the condition table cannot appear in the hybridization table, and vice versa.

Neural network output variables can also appear in the `targetValue` column of the condition table.

Reviewer: > Neural network input variables are valid `targetId`s for the condition table **as long as, following the PEtab standard, they are NON_PARAMETER_TABLE_ID**.

Inputs cannot appear in the parameter table already, right? They need to be mapped to estimated parameters via the hybridization or conditions table. Not sure if this is in the docs yet though.

Reviewer: I see a use case for setting a NN input to be an estimated parameter, e.g. maybe a user wants the NN to take an estimated reaction rate constant as input.

But I don't see a use case for converting a NN input into an estimated parameter... if there is no use case, I think it's safer to restrict the appearance of neural network PEtab IDs in all PEtab tables, to make implementation easier.

Author: > But I don't see a use case for converting a NN input into an estimated parameter... if there is no use case, I think it's safer to restrict the appearance of neural network PEtab IDs in all PEtab tables, to make implementation easier.

I think we have discussed this exact point before :) I do want to avoid restrictions based on the user will not need this, so I think we should allow this, and then let tools choose whether or not to support it (if tricky to implement, I think this is definitely one of the features one does not need to support)

### Intra-simulation hybridization

Intra-simulation neural network models depend on model simulated model quantities (see case 2 [here](@ref hybrid_types)).

#### Inputs

Valid `targetValue` for a neural network input is an expression that depend on model species, time, and/or parameters. Any model species and/or parameters in the expression are expected to be evaluated at the given time-value.

#### Outputs

Valid `targetId` for a neural network output is a constant model parameter. During PEtab problem import, any assigned parameters is replaced by the neural network output in the ODE RHS.

> **Importantly**, it is invalid for a model parameter assigned by a neural network output to appear in both the ODE RHS and the observable formula. In the observable formula, neural network output PEtab identifiers are expected to be directly encoded, making a neural network–assigned parameter in the formula ambiguous. Conversely, it is considered valid if a neural network output is used to assign an ODE model parameter and is also directly encoded in the observable formula.

Reviewer 2: ```suggestion
**Importantly**, it is invalid for a model parameter assigned by a neural network output to appear in both the ODE RHS and the observable formula. In the observable formula, neural network output PEtab identifiers are expected to be directly encoded, making a neural network–assigned parameter in the formula ambiguous. Conversely, it is considered valid if a neural network output is used to assign an ODE model parameter and is also directly encoded in the observable formula.
```

I wouldn't put this in the spec but leave to individual implementations to support or not support this.

Author: This is a fair point, thanks!

  - A numeric value applied to all parameters under `parameterId`.
- `estimate` [0 | 1, REQUIRED]: Indicates whether the parameters are estimated (`1`) or fixed (`0`). This must be consistent across layers. For example, if `netId.parameters` has `estimate = 0`, then potential layer rows must also be `0`. In other words, freezing individual network parameters is not allowed.

### Bounds for neural net parameters

Bounds can be specified for an entire network or its nested identifiers. However, most optimization algorithms used for neural networks, such as ADAM, do not support parameter bounds in their standard implementations. Therefore, neural net bounds are optional and default to `-inf` for the lower bound and `inf` for the upper bound.

## [Problem YAML File](@id YAML_file)

> An extension section is included in the PEtab SciML YAML file for specifying neural network YAML files, as well as array parameter, input, and output files. These elements are defined using the following key-value mappings:
> 
> - `file[extensions][petab_sciml][neural_nets]`: Neural network models. Each network is defined as a key-value mapping, where the key is the unique neural network Id (`netId`), and the corresponding value is another key-value mapping:
>   - `[netId][location]`: The file path where the neural network model is stored.
>   - `[netId][format]`: The neural network format. Expected to be `YAML` if the network is provided in the PEtab SciML library [YAML format](@ref YAML_net_format). Otherwise, the neural network library should be provided (e.g Lux.jl or equinox.py).
>   - `[netId][hybridization]`: The neural network hybridization type. Expected to be either `pre_simulation` or `intra_simulation` (for type information see [here](@ref hybrid_types)).
> - `yaml_file[extensions][petab_sciml][array_files]`  Potential array files. Parameter files are expected to follow the structure described [here](@ref hdf5_ps_structure), and input files should follow the structure described [here](@ref hdf5_input_structure). Each entry is defined with a key-value mapping where the key is the array Id (`arrayId`), and the corresponding value is another key-value mapping:
>   - `[arrayId][location]`: The file path.
>   - `[arrayId][format]`: The file format (e.g., HDF5).

Reviewer 2: ```suggestion
The `petab_sciml` extension is defined within the `extensions` section of a PEtab YAML file. It specifies the configuration of neural networks and optional array files used for simulation or parameter estimation.

### Fields

#### `neural_nets` (required)

A list of neural network definitions. Each entry is a mapping with the following keys:

- **`location`** (`str`):  
  File path to the neural network model.

- **`format`** (`str`):  
  Format of the neural network. Use `YAML` if the model is defined in the [PEtab SciML YAML format](@ref YAML_net_format).  
  For models defined using external libraries, specify the library name (e.g., `Lux.jl`, `equinox.py`).

- **`dynamic`** (`bool`):  
  Indicates the hybridization type:  
  - `true`: dynamic hybridization  
  - `false`: static hybridization  
  For more information, see [Hybrid Types](@ref hybrid_types).

#### `array_files` (optional)

A list of array file definitions. Each entry is a mapping with the following keys:

- **`location`** (`str`):  
  File path to the array file.

- **`format`** (`str`):  
  Format of the file (e.g., `HDF5`).

Parameter array files must follow the structure described in [HDF5 Parameter Structure](@ref hdf5_ps_structure).  
Input array files must follow the structure described in [HDF5 Input Structure](@ref hdf5_input_structure).
```

# Format Specification

A PEtab SciML problem extends the PEtab standard version 2 to accommodate hybrid models SciML problems that combine data-driven (neural net) and mechanistic components. The extension introduces one new PEtab file type:

1. **Neural Net File(s)**: YAML file(s) describing neural net model(s).

It further extends the following standard PEtab files:

1. **Mapping Table**: Used to describe how neural network inputs and outputs map to PEtab quantities.
2. **Parameters Table**: Used to describe nominal values and potential priors for initializing network parameters.
3. **Condition Table**: Used to assign neural network outputs and inputs.
4. **Problem YAML File**: Includes a new SciML field.

All other PEtab files remain unchanged. This page explains the format and options for each file that is added or modified by the PEtab SciML extension.

## High Level Overview

The main goal of the PEtab SciML extension is to enable hybrid models that combine data-driven and mechanistic components. There are three types of hybrid model considered, each specified differently:

> 1. **Data-driven models in the ODE model’s right-hand side (RHS):** In this scenario, the SBML file is modified during import by either replacing a derivative or assigning a parameter to a neural network output. In both cases, the neural network input and output variables (as defined in the mapping table) must be assigned in the condition table using the `setNetRate` and/or `setNetAssignment` operator types.

Author: I added the `Net` designation to simplify identifying when a neural network is assigned in the conditions table, and because I initially thought the PEtab v2 keywords were specified for certain rules. However, after reading the current specification, I realize there is no difference— somehow I had assumed that `setRate` was only allowed for rate-rule expressions. If you prefer using the original PEtab v2 names, I can change this.

On another note, I realized that `setAssignment` raises an interesting semantic point. According to the standard, it should target an algebraic expression. However, when inserting a neural network into the ODE RHS, it is convenient to replace a constant model parameter. This leads to the question: Should `setAssignment` be allowed to assign values to constant parameters, or should we enforce that parameters replaced by neural networks must be nonconstant in the model file?

> 1. **Data-driven models in the ODE model’s right-hand side (RHS):** In this scenario, the SBML file is modified during import by either replacing a derivative or assigning a parameter to a neural network output. In both cases, the neural network input and output variables (as defined in the mapping table) must be assigned in the condition table using the `setNetRate` and/or `setNetAssignment` operator types.

Reviewer 2: > I added the `Net` designation to simplify identifying when a neural network is assigned in the conditions table, and because I initially thought the PEtab v2 keywords were specified for certain rules. However, after reading the current specification, I realize there is no difference— somehow I had assumed that `setRate` was only allowed for rate-rule expressions. If you prefer using the original PEtab v2 names, I can change this.

Yes, would argue it keeps the spec simpler.

> On another note, I realized that `setAssignment` raises an interesting semantic point. According to the standard, it should target an algebraic expression. However, when inserting a neural network into the ODE RHS, it is convenient to replace a constant model parameter. This leads to the question: Should `setAssignment` be allowed to assign values to constant parameters, or should we enforce that parameters replaced by neural networks must be nonconstant in the model file?

Yes, I would stick to that as it makes validation simpler and also allows developers to parse whether the NN output can be evaluated offline (setValue) or has to be done online (setAssignment).

3. **Data-driven models before the ODE model:** In this scenario, the data-driven model sets constant parameters or initial values in the ODE model prior to simulation. The input variable (as defined in the mapping table) can be assigned in the parameter or condition table as a standard constant PEtab variable, and the output variables (as defined in the mapping table) are assigned via the condition table.

## Neural Network Model Format

**TODO:** Dilan, I will need some help from you here, link the scheme?

Neural network models are provided as separate YAML files, and each tool supporting the extension is responsible for importing this file into a suitable format. A neural network YAML file has two main sections:

- **layers**: Defines the neural network layers, each with a unique ID. The layer names and argument syntax follow PyTorch conventions.
- **forward**: Describes the forward pass, specifying the order of layer calls and any applied activation functions.

> Although network YAML files can be manually written, the recommended approach is to define a PyTorch `nn.Module` whose constructor sets up the layers and whose `forward` method specifies how they are invoked. For example, a simple feed-forward network can be defined as:

Author: What does open mean in this context? Would a problem be considered valid if, for example, the neural network model is an Equinox module or a Julia Lux model? I am open to supporting that option, as I can imagine that some more exotic architectures might be difficult to capture in the YAML format.

In that case I agree, we can say that the YAML is specific to the library and is supported across packages, while each package is free to support additional methods for providing the model.

> Although network YAML files can be manually written, the recommended approach is to define a PyTorch `nn.Module` whose constructor sets up the layers and whose `forward` method specifies how they are invoked. For example, a simple feed-forward network can be defined as:

Reviewer 2: > What does open mean in this context? Would a problem be considered valid if, for example, the neural network model is an Equinox module or a Julia Lux model? I am open to supporting that option, as I can imagine that some more exotic architectures might be difficult to capture in the YAML format.
> 
> In that case I agree, we can say that the YAML is specific to the library and is supported across packages, while each package is free to support additional methods for providing the model.

Yes

### Neural Network Input

When network input is provided as an array via an HDF5 file (see the mapping table below), it should follow the PyTorch convention. For example, if the first layer is `Conv2d`, the input should be in `(C, W, H)` format, with data stored in row-major order. In general, the input should be structured to be directly compatible with PyTorch.

!!! tip "For developers: Respect memory order"
    Tools supporting the SciML extension should, for computational efficiency, reorder input data and potential layer parameter arrays to match the memory ordering of the target language. For example, PEtab.jl converts input data to column-major order, as used in Julia.

## Mapping Table

The mapping table describes whcih PEtab problem variables a neural network’s inputs and outputs map. Each neural network input and output must be mapped in the mapping table.

| **petabEntityId** | **modelEntityId** |
|-------------------|-------------------|
| e.g.              |                   |
| k1                | netId.input1      |

> ### Detailed Field Descriptions

Author: Fair point, I basically just copy the spec here. Will remove.

- **petabEntityId [STRING]**: A valid PEtab identifier not defined elsewhere in the PEtab problem. It can be referenced in the condition, measurement, parameter, and observable tables or be a file, but not in the model itself. For neural network outputs, the PEtab identifier must be assigned in the condition table, whereas for inputs, this is not required (see examples below).

> - **modelEntityId [STRING]**: Describes the neural network entity corresponding to the `petabEntityId`. Must follow the format `netId.input{n}` or `netId.output{n}`, where `n` is the specific input or output index.

Reviewer 2: How about using array indexing instead `{netId}.input[{n}]`. Maybe also introduce layer to allow multi-headed/multi-input networks? E.g. `{netId}.layer.input[{n}]`.

> - **modelEntityId [STRING]**: Describes the neural network entity corresponding to the `petabEntityId`. Must follow the format `netId.input{n}` or `netId.output{n}`, where `n` is the specific input or output index.

Author: I think the array indexing looks better!

I also like the option for multi-layer input, but we need to consider how to implement this in practice. With the current approach—defining a YAML file with a layer specification in the init method followed by the forward function—we don't have a way to insert outside layer-specific input in the forward trace. Rather, I think we can specify the input argument names to the forward trace. @dilpath, if I'm not mistaken, the standard allows more than one input argument to the forward function, right? Then, for multi-layer input, it could be something like `{netId}.input.argName{n}[{}]` in the mapping table, and the forward-trace function would be like `forward(argName1, argName2)` where different arguments in the forward function could now target different layers. Moreover, I don't believe that Equinox (or rather Jax) would have any issues with a forward trace that accepts multiple arguments.

> - **modelEntityId [STRING]**: Describes the neural network entity corresponding to the `petabEntityId`. Must follow the format `netId.input{n}` or `netId.output{n}`, where `n` is the specific input or output index.

Reviewer 2: sounds good!

> - **modelEntityId [STRING]**: Describes the neural network entity corresponding to the `petabEntityId`. Must follow the format `netId.input{n}` or `netId.output{n}`, where `n` is the specific input or output index.

Reviewer: Yes, multiple input arguments should work fine. Array indexing looks good :+1: I could then implement this interface in the Python class so this mapping table format matches how one would access the value(s) in Python exactly.

> - **modelEntityId [STRING]**: Describes the neural network entity corresponding to the `petabEntityId`. Must follow the format `netId.input{n}` or `netId.output{n}`, where `n` is the specific input or output index.

Author: Great, then I will change to array indexing. Moreover, I will later add test-cases where we have more than one input argument to the forward function for the data-driven model.

### Network with Scalar Inputs

For networks with scalar inputs, the PEtab entity should be a PEtab variable. For example, assume that the network `net1` has two inputs, then a valid mapping table would be:

| **petabEntityId** | **modelEntityId** |
|-------------------|-------------------|
| net1_input1       | net1.input1       |
| net1_input2       | net1.input2       |

Scalar input variables can then be:

- **Parameters in the parameters table**: These may be either estimated or constant.
- **Parameters assigned in the condition table**: More details on this option can be found in the section on the condition table.

### Network with Array Inputs

> Sometimes, such as with image data, a neural network requires array input. In these cases, the input can be specified as an HDF5 file path directly in the mapping table:

Reviewer 2: I would prefer to introduce a new column instead, as the hdf5 file is not a valid PEtab id. Maybe `fileLocation` and potentially additional columns for groups/fields/attributes? Also use `net1.input` to indicate this assigns an array?

Author: > Also use net1.input to indicate this assigns an array? 

I agree!

As for the other part, you mean something like this:

| **petabEntityId** | **modelEntityId** | **fileLocation** |
|-------------------|-------------------|------------------|
|         | net1.input        |     net1.hdf5              |

And for condition specific inputs something like this:

| **petabEntityId** | **modelEntityId** | **fileLocation** |
|-------------------|-------------------|------------------|
| net1_input        | net1.input        |                  |
| net1_input_cond1  |                   | net1_cond1.hdf5  |
| net1_input_cond2  |                   | net1_cond2.hdf5  |

Where `net1_input_cond1` are variables assigned to set value for `net1_input` in the conditions table?

Anyhow, this comment raises the question of correctness versus ease of specifying the problem. If file locations are specified in the mapping table for inputs, this opens up the possibility of adding extra columns with relevant attributes if needed. However, if there are many conditions, the user must enter numerous entries in both the mapping and condition tables. Nonetheless, I believe that using `fileLocation` might be the better solution, as it offers flexibility for providing additional information when necessary, and is more PEtab correct. @dilpath, what do you think?

Reviewer 2: Yes, that's exactly what I had in mind.

I agree that it may become cumbersome in terms of specification for multiple conditions. Maybe it makes some sense to introduce some conventions about how to directly encode this in hdf via dimension scales or attributes (see, e.g. https://stackoverflow.com/questions/41704019/how-to-use-hdf5-dimension-scales-in-h5py), only assign ones with the mapping over conditions implied?

Author: Yes, that could be a good solution for handling multiple conditions. Overall, I think we can allow assign files in both the condition and mapping tables. 

For many condition models, we might leverage the attribute field in the HDF5 file to allow for a single file. For example, if the same file variable is used across multiple conditions, then in the HDF5 file the input should be accessible as something like `file.cond_id`?

Reviewer: I agree, it's much nicer to keep all the rest of the PEtab files "valid" (at least in terms of IDs) by the v2 spec, and only extend the mapping table. Instead of `fileLocation`, how about `scimlNominalValue`? Since this will be ignored if the layer parameters are set to be estimated, right? It makes it clear that it's an extension column, that can be interpreted similarly to the parameter table `nominalValue` column.

Doing everything via the mapping table might be conceptually more complicated for users though, than being able to write e.g. `netId1.layer1` directly in the parameters table. Perhaps we could get some other opinions at the next PEtab editor meeting, if you want to present this then?

Reviewer: For the condition table, something like this?

Mapping table

| **petabEntityId** | **modelEntityId** | **scimlNominalValue** |
|-------------------|-------------------|------------------|
|     net1_input    | net1.input        | net1_input.hdf5                 |
| net1_input_cond2  |                   | net1_input_cond2.hdf5  |

Condition table

| **conditionId** | **operatorType** |  **targetId** |**targetValue** |
| --- | --- | --- | --- |
| cond2 | setValue | net1_input | net1_input_cond2 |

This provides default nominal values for `net1.input`, and condition-specific values for `cond2`.

Author: > Doing everything via the mapping table might be conceptually more complicated for users though, than being able to write e.g. netId1.layer1 directly in the parameters table. 

I agree on this point, however, I think we can work around this with having clear tutorials. Moreover, via the parameters table users can already assign for the entire network model via `netId`, so I do not know how common it will be to target specific layers. But yes, we can also discuss this a bit at the next editors meeting (I can present the spec which I should have updated to the comments herein)

>  Instead of `fileLocation`, how about `scimlNominalValue`

Hmm, I think I still prefer `fileLocation`. `scimlNominalValue` definitely makes the most sense for nominal values, however, might be confusing when referring to inputs, and I would like to not introduce to many columns. @FFroehlich  what do you think? 

> For the condition table, something like this?

Yes, this is how I imagine it to look like!

Reviewer: > > Instead of `fileLocation`, how about `scimlNominalValue`
> 
> Hmm, I think I still prefer `fileLocation`. `scimlNominalValue` definitely makes the most sense for nominal values, however, might be confusing when referring to inputs, and I would like to not introduce to many columns. @FFroehlich what do you think?

Yes you're right about inputs... I guess the idea is to assign inputs a valid PEtab ID in the mapping table, and then specify them as inputs in the conditions table.

I guess we will introduce the format as e.g. the "SciML HDF5 format for input and parameter specification". So we could also name the column `scimlHdf5` so it explains itself.

mkstd can be used to generate this custom HDF5 format too -- I can make a suggestion but if you have an idea of what the structure could be, let me know.

> > For the condition table, something like this?
> 
> Yes, this is how I imagine it to look like!

Nice, then I think there is no issue for multiple conditions and nothing needs to be encoded in the HDF5, or am I missing something?

Author: > I guess we will introduce the format as e.g. the "SciML HDF5 format for input and parameter specification". So we could also name the column scimlHdf5 so it explains itself.

This would be good! As for the structure, nothing really complicated, for parameters something like:

```
parameters.hdf5
└───layer1
│   ├── arrayId{1}
│   └── arrayId{2}
└───layer2
    ├── arrayId{1}
    └── arrayId{2}
```

and for input 

```
input.hdf5
└───input
    └─── inputArray
```
So do not know if we need a custom format?

> Nice, then I think there is no issue for multiple conditions and nothing needs to be encoded in the HDF5, or am I missing something?

It is more that if there are many conditions (>20) it gets cumbersome to encode a condition hdf5-file for each, so would be nice to have this in the conditions table.

Reviewer 2: Yes, I think it would be nice to introduce some implicit mapping between PEtab data structures and the HDF5 file format to make specification a bit more convenient and we don't end up with too many files/expansive mapping.

I would prefer `fileLocation` over `scimlHdf5` as this makes everything a bit more generalisable. For example, one could also think about allowing csv files.

Reviewer: I agree but this makes me wonder, was there a good reason to not put array files in the PEtab YAML like all other files in PEtab normally? e.g.
```yaml
extensions:
  petab_sciml:
    array_files:
      - a_layer.hdf5
      - another_layer.hdf5
```

The v2 `model_files` key in the YAML also supports specification of ID (don't need to use mapping table) and explicit specification of file type. Why not re-use this format for array files? Also the net files are currently defined with a "file" key, but we could use "location" like for v2 `model_files`.
```yaml
  model_files:
    model_0:
      language: sbml
      location: _model.xml
```

Since we already support specification of layer IDs in the MLModel YAML file, why not use those IDs instead of defining things in the mapping table?
https://github.com/sebapersson/petab_sciml/blob/f7c37800cc7f4ee58e6eb55842bbdbc6e6eb6427/src/python/petab_sciml/specification.py#L17-L34

This approach would mean there is no need for the mapping table (and condition information in the HDF5 would still be possible).

Author: This is a valid point; however, including files directly in the YAML file reintroduces the valid PEtab ID problem. According to the PEtab standard, a PEtab ID cannot be a file path. Therefore, referring to an array file in the conditions or parameter table is not valid. So with the YAML file, users would first define the array in the YAML file and then map it to a PEtab ID in the mapping table. Therefore, I think it is easier to only do the mapping the in mapping table once.

Reviewer: I agree, not simpler, but in terms of consistency, once a user understands how to specify an SBML file in the PEtab YAML, then they will understand how to specify array files too, without having to learn the mapping table format.
So "`array_files` like `model_files`" suggestion is
```yaml
extensions:
  petab_sciml:
    array_files:
        myArray1:
            language: hdf5
            location: a_layer.hdf5
        myArray2:
            language: tsv
            location: another_layer.tsv
```

So I think there is a benefit for people getting started with the format.

I think the counterpoint would be on heavy usage usability -- a mapping table is probably easier to edit. But maybe not too much worse for YAML.
- Manually editing: table easier because e.g. can use Excel instead of a YAML code editor, harder to make a formatting mistake with tables
- Programmatic-editing independent of PEtab: both simple as long as you have a YAML library that gives you a dictionary to edit

Author: Hmm, personally I often try to prioritize user friendliness, for which I think the mapping table is nicer. Lets get a third opinion, @FFroehlich

Reviewer 2: I would argue that you achieve user friendliness through modularity, i.e. not too much interaction between individual files/tables, files specifying as few operations as possible and consistently specifying the same types of operations in the same location. I believe all three by putting the file specification in the yaml file.

Also simplifies problem loading as we don't have to keep track of local paths when interpreting any table.

Author: > Also simplifies problem loading as we don't have to keep track of local paths when interpreting any table.

Good point. 

Ok, I can then agree that going for the YAML file is the way to go. I will update the spec with respect to this, and I will also add in the spec how to add condition information in an input file.

Reviewer: > I will also add in the spec how to add condition information in an input file.

Can leave this out for now if you wish (or simply write that it will be possible) -- I can first make a suggestion for the HDF5 format and where this information goes as a separate PR where we can discuss it, before documenting it

Author: Yes, I can leave it out (also makes it a bit easier to update the spec :)

| **petabEntityId**   | **modelEntityId**  |
|-------------------|------------------|
| input_net1.hdf5    | net1.input1      |

As mentioned in [ADD], the HDF5 file should follow PyTorch indexing and be stored in row-major order.

When there are multiple simulation conditions that each require a different neural network array input, the mapping table should map to a PEtab variable (e.g., `net1_input`):

| **petabEntityId** | **modelEntityId**  |
|-----------------|------------------|
| net1_input      | net1.input1      |

> This variable is then assigned to specific input files via the condition table using the `setValue` `operatorType`. For a full example of a valid PEtab problem with array inputs, see [ADD].

Reviewer 2: Potentially keep files out of the condition table and rather assign data entries to petab ids in the mapping table instead?

Author: See comment above.

> ### Network Observable Formula Output

Reviewer 2: Do we need this section? This should be evident from the PEtab spec

Author: Agree. Anyhow, I will probably link an example here (from test-suite), but will remove the rest.

### Additional Details

Although a neural network can, in principle, accept both array and scalar inputs, this feature is not currently tested for among tools implementing the PEtab SciML extension due to it being hard to implement. However, tools are free to add this feature.

## Condition Table

In the PEtab SciML extension, the condition table is extended to specify how neural network outputs (and, if necessary, inputs) are assigned. Two new `operatorType` are introduced in the extension to support this functionality:

1. **setNetRate**: Assigns the rate of a species to a neural network output. Here, `targetValue` must be a neural network output and `targetId` must be a model specie.
2. **setNetAssignment**: Assigns the input or output of a neural network in the ODE right-hand side (RHS) or the input in the observable formula.
   - Input Case: `targetId` is a neural network input, and `targetValue` can be any valid PEtab math expression that references model variables.
   - Output Case: `targetId` is a non-estimated ODE model parameter, and `targetValue` is a neural network output. This is used to assign a neural network output in the ODE model RHS.

> !!! warn "Model structure altering conditions"

Reviewer 2: Would follow same rules as petab spec here (differentiable/algebraic targets).

Author: As this is crucial for problem formulation, I think it is good to be repeating stuff so users do not have to jump between documentations. But maybe add something like:

follow the PEtab spec here on (differentiable/algebraic targets), which means...

- If no inputs use `setNetAssignment` and all outputs use either `setValue` or `setInitial`, the neural network sets model parameters or initial values before the simulation.

All other combinations are disallowed because they generally do not make sense in a PEtab context. For example, if inputs use `setNetAssignment` and outputs use `setValue`, parameter values prior to simulation would be set via an assignment rule consisting of model equations, which is not permitted in PEtab as assignment rules might be time-dependent. Moreover, if a parameter is to be set via an assignment rule, this should already be coded in the model. Implementations must ensure that the input combinations in the condition table are valid.

## Parameter Table

The parameter table follows the same format as in PEtab version 2, with a subset of fields extended to accommodate neural network parameters and new `initializationPriorType` values for neural network-specific initialization. A general overview of the parameter table is available in the PEtab documentation; here, the focus is on extensions relevant to the SciML extension.

### Detailed Field Description

> - **parameterId [String]**: Identifies the neural network or a specific layer/parameter array. For example, `layerId` for `netId` can be specified using `netId.layerId`. A row for `netId` must be defined in the table. When parsing, more specific levels (e.g., `netId.layerId`) take precedence for nominal values, priors, etc.

Reviewer 2: Would prefer to stick to PEtab ids here and assign them in the mapping table.

Author: Fair, if we agree upon this for input files then for consistency the same should hold for the parameters table.

Then the edge case, is a `netId` a valid PEtab variable (I think it should be)?

Reviewer 2: > Then the edge case, is a `netId` a valid PEtab variable (I think it should be)?

Yes, but as ids should be unique and `netId` already refers to the network itself, I would probably refer to the parameters via `netId_p` or `netId_weights` instead.

Reviewer: Agreed... each layer has inputs, outputs, and trainable parameters. I guess most people will think `netId` alone represents the output, though if it appears in the parameters table then it's also clear it means the parameters. To avoid this interpretation, we could require all instances of `netId1` or `netId1.layer[n]` alone to be invalid, and require explicit reference to e.g. `netId1.inputs[n]`, `netId1.parameters[n]`, `netId1.outputs[n]` (indexing optional if referring to a specific subset), similarly `netId1.layer[n].parameters[n]`. Shortcuts like `netId1.layer[n].weights[n]` are still fine for me (other options: `netId1.layer[n].parameters.weights[n]`, `netId1.layer[n].parameters[weights]` to highlight that weights are some indexing/subset of parameters).

So I would say `netId` is not a valid PEtab variable (and should be explicitly forbidden to avoid context-dependent interpretations). Only valid in the SciML extension section of the PEtab YAML file.

Author: > So I would say `netId` is not a valid PEtab variable (and should be explicitly forbidden to avoid context-dependent interpretations). Only valid in the SciML extension section of the PEtab YAML file.

I agree, I will update this in the specs and for parameters use something like `netId.parameters` (`parameters` is definitely the most clear think here). As a consequence, this means that  `netId.parameters` will have to be mapped in the mapping table to get a valid PEtabId for the parameters table, but as this only happens once per neural network model, I am fine with it.

Lastly, I prefer to forbid value indexing for parameters (e.g., `netId1.layerId.weights[n]`). Allowing such fine-grained assignment would require users to work with PyTorch indexing, which might not align with indexing in the language they use, and further I believe working at the array level is sufficient for these models.

Reviewer: > Lastly, I prefer to forbid value indexing for parameters (e.g., `netId1.layerId.weights[n]`). Allowing such fine-grained assignment would require users to work with PyTorch indexing, which might not align with indexing in the language they use, and further I believe working at the array level is sufficient for these models.

Makes sense :+1:

Reviewer: > Lastly, I prefer to forbid value indexing for parameters (e.g., `netId1.layerId.weights[n]`). Allowing such fine-grained assignment would require users to work with PyTorch indexing, which might not align with indexing in the language they use, and further I believe working at the array level is sufficient for these models.

Currently the docs use indexing for individual input/output scalars (e.g. `net1.input[1]`). So, is 2D image data possible, or would this require some decision on row-major or column-major matrices?

Author: Good question. Basically, I do not want to allow individual indexing on parameters for the reasons above.

For input and output, I think indexing should be acceptable as long as we are dealing with vector inputs/outputs. Then, if we need 2D image data input, then the input should be provided via a file. In the HDF5 file format, we can accept any kind of input array, assuming the input is compatible with the architecture and is adapted according to the library used. Also, there is a lot of flexibility with having more input arguments. Lastly, I do not see any application where larger than 1D array would make sense for the output. 

If this sounds good, I can clarify this in the documentation.

Reviewer: > assuming the input is compatible with the architecture and is adapted according to the library used

I would make this assumption explicit in the YAML (e.g. `array_order: row-major`), otherwise problems are not interoperable I think... A tool can determine whether it supports the problem by checking the `array_order`.

And if the assumption is explicit, then multi-dim indexing would be completely fine, since we could write a tool to convert a PEtab SciML problem from one array order to the other (when that becomes useful...).

Author: > And if the assumption is explicit, then multi-dim indexing would be completely fine, since we could write a tool to convert a PEtab SciML problem from one array order to the other (when that becomes useful...).

Different languages use different array indexing conventions. For example, index[1, 2, 3] in PyTorch is represented as [3, 2, 1] in Lux.jl for Julia. So if a user writes `ps[3, 2, 1]` in some table, what would it mean (and if the user has to specify what it means, we are ending up with many required inputs from the user, like array_order, indexing_system, format...).

Therefore, I suggest avoiding array indexing entirely and leaving it to the tools in combination with the users to ensure that the HDF5 files are compatible with the chosen architecture. If the YAML format is used, the PEtab-SciML library can help verify inputs have correct shape. If a tool uses another neural network library (for example, allowing users to write the network in Lux.jl), that tool should generate the correctly indexed parameter files, and functions to check that input is compatible with the forward functions.

Reviewer: > we are ending up with many required inputs from the user

Agreed, if it gets to this point, we should combine them into some `framework: julia/equinox/pytorch/...` field. Anyway, fine with me if we leave out multi-dim indexing while there is no important use case that requires it.

So, putting that aside for now, we still need to ensure interoperability. HDF5 is row-major. I assume you'll use some Julia HDF5 library that handles this for you, such that you see column-major arrays in Julia, but it's always row-major arrays on disk. If this is not the case, then we will need something (HDF5 used to have a `perm` field but that is deprecated/removed...).

Author: > I assume you'll use some Julia HDF5 library that handles this for you, such that you see column-major arrays in Julia,

The HDF5.jl library manages this in Julia. In general, I think this is handled by most importing YAML libraries across languages.

Taken together, there should be no issues with inter-operability if the user employs the YAML standard model format (we also have a robust test suite for this purpose so implementers can test correct import). Further, if the user provides the neural network model in a specific programming language (for example, as an `Equinox.Module`), then—according to the documentation—the HDF5 files should be stored in a way that is directly compatible with that library, and no inter-language interoperability is expected.

Reviewer: Great, thanks, then that clarifies things for me.

> (for example, as an `Equinox.Module`), then—according to the documentation—the HDF5 files should be stored in a way that is directly compatible with that library

This is still unclear to me... frameworks will presumably be able to read from, or write to, the YAML format -- if not, what's the point of a framework supporting PEtab SciML (since currently most of the SciML info is encoded in our YAML and HDF5 formats)? The framework may as well use their own format and live in their own bubble, since they gain nothing from supporting PEtab SciML without the YAML (and/or HDF5)? But this discussion can be left for when we have examples of this framework-specific specification.

Author: > This is still unclear to me... frameworks will presumably be able to read from, or write to, the YAML format -- if not, what's the point of a framework supporting PEtab SciML (since currently most of the SciML info is encoded in our YAML and HDF5 formats)? The framework may as well use their own format and live in their own bubble, since they gain nothing from supporting PEtab SciML without the YAML (and/or HDF5)? But this discussion can be left for when we have examples of this framework-specific specification.

This approach allows for more complex architectures that may not be fully supported by the YAML standard (e.g., transformers). In such cases, users will provide the model in the framework where the YAML is imported (e.g., Lux.jl), while still using the rest of the format to specify inputs, outputs, simulation conditions, etc. (hence the need of the standard)

Of course, having a standardized YAML format across different programming languages is preferable, so it remains the recommended approach for users, and also what the test-suite is built upon.

Reviewer: Yes good point, this is suboptimal but then we should aim to increase the scope of the YAML format to support popular models like transformers over time...

- Should we expect that all tools support the YAML format where it's sufficient, and only use their own formats where it's insufficient? Or really that some frameworks don't support the YAML at all?
- Should we create a common "ScimlModel" class that defines the interface required for customs models to be compatible with the framework, such that e.g. the PEtab SciML library can easily extract information like layers and parameters from any custom model? This would enable at least some interoperability, despite custom formats.

Author: > Should we expect that all tools support the YAML format where it's sufficient, and only use their own formats where it's insufficient? Or really that some frameworks don't support the YAML at all?

I think tools should support the YAML, so they can use the YAML to test against the test-suite. Also, once a tools supports the YAML model, it should be straightforward for a tool to add support for the neural network library they use.

> Should we create a common "ScimlModel" class that defines the interface required for customs models to be compatible with the framework, such that e.g. the PEtab SciML library can easily extract information like layers and parameters from any custom model? This would enable at least some interoperability, despite custom formats.

Sounds good, but I think we can look at this a bit in the future when we have everything in place.

> Yes good point, this is suboptimal but then we should aim to increase the scope of the YAML format to support popular models like transformers over time...

Agree, but meanwhile I think it is nice to avoid any potential problems where user complains about their specific architecture not being possible.

Reviewer: Great, sounds good to me too, thanks for explaining

> - **nominalValue [String \| NUMERIC]**: Specifies neural network nominal values. This can be:

Reviewer 2: Same here, keep those files in the mapping table

| net1.layer1   | lin              | -inf         | inf          | 1          | net1_ps.hf5    | kaimingUniform            | 1                               |
| net1.layer2   | lin              | -inf         | inf          | 1          | net1_ps.hf5    | kaimingUniform            | 5/3                             |

If is also possible to have different priors for different parameter arrays in a layer. For example, to use different priors for the weights and bias in `layer1` of `net1`, which is, for instance, a `linear` layer. In this case, a valid parameter table would be:

| **parameterId** | **parameterScale** | **lowerBound** | **upperBound** | **estimate** | **nominalValue** | **initializationPriorType** | **initializationPriorParameters** |
|--------------------|------------------|--------------|--------------|------------|----------------|---------------------------|---------------------------------|
| net1               | lin              | -inf         | inf          | 1          | net1_ps.hf5    | kaimingUniform            | 1                               |

> | net1.layer1.weight | lin              | -inf         | inf          | 1          | net1_ps.hf5    | kaimingUniform            | 1                               |

Reviewer 2: This uses a single file, needs to be specified how groups/fields/attributes are mapped to the net.

Author: Thanks, forgot to add this. Basically, the single file describes all parameters for a neural network, so if the same file is refereed to, it takes the nominal values from said file.

Reviewer 2: right, but we need to specify how the hdf5 data structure is mapped to the neural network

| net1.layer1.bias   | lin              | -inf         | inf          | 1          | net1_ps.hf5    | kaimingNormal             | 5/3                             |

### Bounds for neural net parameters

Bounds can be specified for an entire network or its nested levels. However, it should be noted that most optimization algorithms used for neural networks, such as ADAM, do not support parameter bounds in their standard implementation.

## Problem YAML File

The PEtab problem YAML file follows the format of PEtab version 2, except that a mapping table is required (it is optional in the standard). It also includes an extension section to specify the neural network YAML files:

```yaml
extensions:
  petab_sciml:
    netId1:

>       file: "file_path1.yaml"

Reviewer: These quotation marks are unnecessary, but fine as is.

# Format Specification

A PEtab SciML problem extends the PEtab standard version 2 to accommodate hybrid models (SciML problems) that combine data-driven (neural net) and mechanistic components. The extension introduces one new PEtab file type:

1. **Neural Net File(s)**: Optional YAML file(s) describing neural net model(s).

It further extends the following standard PEtab files:

1. **Mapping Table**: Used to describe how neural network inputs and outputs map to PEtab quantities.
2. **Parameters Table**: Used to describe nominal values and potential priors for initializing network parameters.
3. **Condition Table**: Used to assign neural network outputs and inputs.
4. **Problem YAML File**: Includes a new SciML field for neural network models and potential arrays.

All other PEtab files remain unchanged. This page explains the format and options for each file that is added or modified by the PEtab SciML extension.

## High Level Overview

> The main goal of the PEtab SciML extension is to enable hybrid models that combine data-driven and mechanistic components. There are three types of hybrid model considered, each specified differently:

Reviewer 2: In principle, we also allow combinations of everything below.

Author: We actually test for all combinations, so I will clarify that combinations are supported.

> 1. **Data-driven models in the ODE model’s right-hand side (RHS):** In this scenario, the SBML file is modified during import by either replacing a derivative or setting a parameter value to a neural network output. In both cases, the neural network input and output variables (as defined in the mapping table) must be assigned in the condition table using the `setNetRate` and/or `setNetAssignment` operator types.

Author: According to the standard (at least what I gathered from discussions), `setAssignment` is basically for setting the value of an SBML assignment rule variable. However, when inserting a neural network into the ODE RHS, it is often convenient to replace a constant model parameter, which is PEtab v2 spec should not be done with  `setAssignment`. Similar, `setRate `seems to be targeted towards rate-rule variables.

Therefore I think it is better to have `setNetAssignment` to have an assignment where we allow the input to for example be assigned an algebraic expression (that is not an assignment rule variable). Moreover, I think it makes it more explicit what is neural network model targets, and that with these assignments we are in a sense altering the model structure.

> 1. **Data-driven models in the ODE model’s right-hand side (RHS):** In this scenario, the SBML file is modified during import by either replacing a derivative or setting a parameter value to a neural network output. In both cases, the neural network input and output variables (as defined in the mapping table) must be assigned in the condition table using the `setNetRate` and/or `setNetAssignment` operator types.

Reviewer 2: > According to the standard (at least what I gathered from discussions), `setAssignment` is basically for setting the value of an SBML assignment rule variable. However, when inserting a neural network into the ODE RHS, it is often convenient to replace a constant model parameter, which is PEtab v2 spec should not be done with `setAssignment`.

That's essentially saying that we don't like the way this is handled in the PEtab v2 core spec, I don't think this will fly well especially since there is the requirement in the PEtab spec "PEtab extensions should extend PEtab core with new orthogonal features or tasks, i.e., they should not make trivial changes to PEtab core."

> Therefore I think it is better to have `setNetAssignment` to have an assignment where we allow the input to for example be assigned an algebraic expression (that is not an assignment rule variable).

How about we introduce a `setParameter` operator instead? I think it's a bit more verbose about the intents and purposes and we can argue that this makes sense to add this complexity in this extension as we are anyways altering the model structure.

> Moreover, I think it makes it more explicit what is neural network model targets, and that with these assignments we are in a sense altering the model structure.

I don't think users care about that, that sounds more relevant to tool developers and we can always provide utility functions for that in a library.

> 1. **Data-driven models in the ODE model’s right-hand side (RHS):** In this scenario, the SBML file is modified during import by either replacing a derivative or setting a parameter value to a neural network output. In both cases, the neural network input and output variables (as defined in the mapping table) must be assigned in the condition table using the `setNetRate` and/or `setNetAssignment` operator types.

Author: > How about we introduce a `setParameter` operator instead? I think it's a bit more verbose about the intents and purposes and we can argue that this makes sense to add this complexity in this extension as we are anyways altering the model structure.

I think this might be the solution! In this case we can use `setAssignment` to assign neural-network input in the RHS (arguably neural-network input in the RHS, but also the observable formula is a algebraic target), and use `setParameter` to assign output. The nice thing with this is that we also force the user to replace a parameter in this operation (making it clear that species should not be replaced).

2. **Data-driven models in the observable function:** In this scenario, the neural network output variable (as defined in the mapping table) is directly inserted in the observable formula. Meanwhile, the input variables (also defined in the mapping table) are assigned in the condition table using the `setNetAssignment` operator type.

> 3. **Data-driven models before the ODE model:** In this scenario, the data-driven model sets constant parameters or initial values in the ODE model prior to simulation. The input variable (as defined in the mapping table) can be assigned in the parameter or condition table as a standard constant PEtab variable, and the output variables (as defined in the mapping table) are assigned via the condition table.

Reviewer 2: ```suggestion
3. **Data-driven models to parametrize ODEs:** In this scenario, the data-driven model sets constant parameters or initial values in the ODE model prior to simulation. The input variable (as defined in the mapping table) can be assigned in the parameter or condition table as a standard constant PEtab variable, and the output variables (as defined in the mapping table) are assigned via the condition table.
```

!!! tip "For developers: Respect memory order"
    Tools supporting the SciML extension should, for computational efficiency, reorder input data and potential layer parameter arrays to match the memory ordering of the target language. For example, PEtab.jl converts input data to column-major order, with Julia indexing.

TODO: We will fix condition specific input in the YAML file later.

## Mapping Table

To avoid confusion regarding what a neural network ID (`netId`) refers to (e.g., parameters, inputs, etc.), `netId` is not considered a valid PEtab identifier. Consequently, every neural network input, parameter, and output must be explicitly mapped in the mapping table to a PEtab variable. In the context of the PEtab SciML extension, the relevant mapping table columns are:

- **petabEntityId [STRING]**: A valid PEtab identifier that is not defined elsewhere in the PEtab problem. This identifier can be referenced in the condition, measurement, parameter, and observable tables, but not within the model itself. For neural network outputs, the PEtab identifier must be assigned in the condition table; for inputs, this is not required (see examples below).

> - **modelEntityId [STRING]**: Describes the neural network entity corresponding to the `petabEntityId`. This must specify a parameter set (e.g. `netId.parameters`), an input (e.g. `netId.input[{n}]`), or an output (`netId.output[{n}]`), where `n` is the specific input or output index.

Reviewer 2: I would prefer to refer to different input/output arguments by their name (e.g. name of the variable in the pytorch implementation) rather than some index and only use brackets notation for array indexing.

I would also add details about how to reference parameters/weights in individual layers here.

Author: > I would prefer to refer to different input/output arguments by their name (e.g. name of the variable in the pytorch implementation) rather than some index and only use brackets notation for array indexing.

I do not think I follow this comment, so if we have something like

```python
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.layer1 = nn.Linear(in_features=2, out_features=5)
        self.layer2 = nn.Linear(in_features=5, out_features=5)
        self.layer3 = nn.Linear(in_features=5, out_features=1)

    def forward(self, x):
        x = self.layer1(x)
        x = F.tanh(x)
        x = self.layer2(x)
        x = F.tanh(x)
        x = self.layer3(x)
        return x
```

Then both the input and output name is `x`, and having something like `netId.x` in the mapping table would be confusing. As we also state that the YAML format accepts a valid PyTorch module on this form, it feels a bit harsh to enforce different input and output names?

Author: > I would also add details about how to reference parameters/weights in individual layers here.

Good point!

Reviewer 2: > > I would prefer to refer to different input/output arguments by their name (e.g. name of the variable in the pytorch implementation) rather than some index and only use brackets notation for array indexing.
> 
> I do not think I follow this comment, so if we have something like
> 
> ```python
> class Net(nn.Module):
>     def __init__(self):
>         super(Net, self).__init__()
>         self.layer1 = nn.Linear(in_features=2, out_features=5)
>         self.layer2 = nn.Linear(in_features=5, out_features=5)
>         self.layer3 = nn.Linear(in_features=5, out_features=1)
> 
>     def forward(self, x):
>         x = self.layer1(x)
>         x = F.tanh(x)
>         x = self.layer2(x)
>         x = F.tanh(x)
>         x = self.layer3(x)
>         return x
> ```
> 
> Then both the input and output name is `x`, and having something like `netId.x` in the mapping table would be confusing. As we also state that the YAML format accepts a valid PyTorch module on this form, it feels a bit harsh to enforce different input and output names?

ah that's a good point. can we do something like `net.inputs[i][j]` instead?

Author: > ah that's a good point. can we do something `like net.inputs[i][j]` instead?

Sounds good with me!

### Network with Scalar Inputs

For networks with scalar inputs, the PEtab entity should be a PEtab variable. For example, assume that the network `net1` has two inputs, then a valid mapping table would be:

| **petabEntityId** | **modelEntityId** |
|-------------------|-------------------|
| net1\_input1       | net1.input[1]       |
| net1\_input2       | net1.input[2]       |

Scalar input variables can then be:

- **Parameters in the parameters table**: These may be either estimated or constant.
- **Parameters assigned in the condition table**: More details on this option can be found in the section on the condition table.

### Network with Array Inputs

> Sometimes, such as with image data, a neural network requires array input. In these situations, the input should be specified as an HDF5 file (for file structure [here](@ref hdf5_input_structure)), which is mapped to a suitable PEtab ID in the [YAML file](@ref YAML_file). This identifier variable is then mapped to a PEtab variable in the mapping table. For example, given that `net1` has a file input, a valid mapping table would be:

Reviewer 2: I think we should only map between IDs in the mapping table and then use condition table to assign arrays.

Author: As discussed above array-files should be defined in the YAML file, where they are defined via a key-value pair:

```yaml
extensions:
    array_files:
      netId1_input:
        location: netId1_input.hdf5
        language: hdf5        
      netId1_ps:
        location: netId1_ps.hdf5
        language: hdf5      
``` 

Here I think `netId1_input` should be a valid PEtab Id that we can then assign to an input in the condition table, because also having to map this Id via the mapping table feels cumbersome for the user.

Maybe my wording is off here, and I should rather write:

*In these situations, the input should be specified as an HDF5 file (for file structure [here](@ref hdf5_input_structure)), which is given to a suitable PEtab ID in the [YAML file](@ref YAML_file).*

Note, I changed mapped to given above.

Reviewer 2: yes, fully agree. To me it sounded like the value of the array was directly assigned in the mapping table. Maybe it becomes clearer if a corresponding condition table is also included in the example.

Author: I will try to add the condition table to see how it turns out.

### Network Scalar Output

If the output does not appear in the observable formula, the output variable should still be defined in the mapping table:

| **petabEntityId** | **modelEntityId**  |
|-----------------|-----------------|
| net1\_output1    | net1.output[1]    |

The output parameter (`net1_output1`) is then assigned in the condition table (see below).

### [Network Parameter Values](@id mapping_ps)

The PEtab ID representing the parameters for a neural network model must also be assigned in the mapping table. For example, if the network is called `net1`, a valid mapping table entry would be:

| **petabEntityId** | **modelEntityId**   |
|-------------------|---------------------|
| net1_ps           | net1.parameters     |

Next, `net1_ps` should be assigned properties in the [parameter table](@ref parameter_table).

It is also possible to target a specific layer in the parameter table. To do so, the layer must first be mapped to a PEtab variable. For the case above, a valid mapping table to target `layer1` would be:

| **petabEntityId** | **modelEntityId**         |
|-------------------|---------------------------|
| net1\_ps           | net1.parameters          |

> | net1\_ps\_layer1    | net1.parameters.layer1   |

Reviewer 2: I don't think we should allow both global parameter and layer wise assignment for the same network

Author: I think we should. If a network has many layers, and we only want to have a special initialization of one layer, would it not get cumbersome to have to then assign every layer?

@dilpath what do you think?

Author: It is already in the spec, but I can make it a bit more clear.

### Assigning Neural Network Output

To set a **constant model parameter value** before model simulations, the `setValue` operator type should be used. For example, if parameter `p` is determined by `net1_output1` (mapped to a neural network output in the mapping table), a valid condition table entry is:

| **conditionId** | **operatorType** | **targetId** | **targetValue** |
|-----------------|------------------|--------------|-----------------|
| cond1           | setValue         | p            | net1\_output1    |

Note that this specification allows for condition-specific assignments. For example, `net1_output1` could target a different parameter in another condition or multiple model parameters in the same condition.

To set an **initial value**, the `setInitial` operator type should be used. For example, if the initial value of species `X` comes from `net1_output1`, a valid condition table entry is:

| **conditionId** | **operatorType** | **targetId** | **targetValue** |
|-----------------|------------------|--------------|-----------------|
| cond1           | setInitial       | X            | net1_output1    |

> To set a **model derivative**, the `setNetRate` operator type should be used. For example, if the rate of species `X` is given by `net1_output1`, a valid condition table entry is:

Reviewer 2: what's the difference between `setNetRate` and `setRate`?

Author: See comment above, but briefly, as I understood `setRate` can only act on rate-rule variables, while `setNetRate` is valid for any model derivative (as we are mutating the model when inserting a neural network model)

Author: > Differential Targets: Entities that are defined in terms of a time-derivative, e.g., the targets of SBML rate rules or species that change by participating in reactions (educts or products).

Fair point. Then I will use `setRate `instead of `setNetRate`. But I still think `setNetAssignment `should be kept?

Reviewer 2: see above

| **conditionId** | **operatorType** | **targetId** | **targetValue** |
|-----------------|------------------|--------------|-----------------|
| cond1           | setNetRate       | X            | net1\_output1    |

> To alter the **ODE RHS**, the `setNetAssignment` operator type should be used. For example, if an ODE model parameter `p` should be given by `net1_output1`, a valid condition table entry is:

Reviewer 2: this could also take effect elsewhere, depending on where the variable occurs, right? So not just RHS. Also, what's the difference between `setNetAssignment` and `setAssignment`?

Reviewer 2: That feels a bit hacky, I would prefer to add a library function that checks whether a net operates on the RHS or not.

Author: That can be added, and in that case said function also needs to check that the neural net output target does not enter into the noise and observable formulas.

|-----------------|---------------------|--------------|-----------------|
| cond1           | setNetAssignment    | p            | net1\_output1    |

### Assigning Neural Network Input

When a neural network sets a **constant model parameter value or initial value**, its input variable (as specified in the mapping table) is a standard PEtab variable. If that input variable is not defined in the parameter table, it should be assigned using the `setValue` operator type.

When a neural network sets **a model derivative or alters the ODE RHS**, the input typically depends on model entities. Therefore, the input variable should be assigned using the `setNetAssignment` operator. For example, if neural network input `net1_input1` is given by specie `X`, a valid condition table is:

| **conditionId** | **operatorType** | **targetId** | **targetValue** |
|--------------|------------------|-------------|---------------|
| cond1        | setNetAssignment | net_input1  | X             |

Note, to ensure correct mapping, `setNetAssignment` must always be used for inputs when the neural network is part of the ODE RHS or observable formula.

> ### `operatorType` Defines Hybrid Model Type

Reviewer 2: I would rather move these explanations up before we start listing examples.

TODO: We will fix condition specific input in the YAML file later.

## [Mapping Table](@id mapping_table)

Each neural network is assigned an Id in the PEtab SciML extension section of the PEtab problem [YAML](@ref YAML_file) file. To avoid confusion about what a neural network Id refers to (e.g., parameters, inputs, outputs), a neural network Id is not a valid PEtab identifier. Consequently, every neural network input, parameter, and output must be explicitly imported into the PEtab problem via the mapping table. Repeating parts of the PEtab standard, the key columns in the mapping table are:

- **petabEntityId** [STRING, required]: A valid PEtab identifier is one that is not defined elsewhere in the PEtab problem. It can be referenced in the condition, hybridization, measurement, parameter, observable tables, and in the PEtab problem YAML file (if the variable refers to an array file), but not within the mechanistic model itself.
- **modelEntityId** [STRING, required]: The neural network component that is mapped to the `petabEntityId`. For a neural network with Id `netId`, the valid identifiers are:

>   - `netId.parameters`: Parameters for a neural network model. A specific layer can be referenced with `netId.layerId.parameters`, and specific arrays in a layer can be referenced with `netId.layerId.arrayId`. For parameter arrays, individual indexing is not allowed.

Author: Unless this is really wanted this can create confusion, because then the individual indexing will depend on the neural network language. For example, for YAML format it should be PyTorch. For Lux.jl network it should be Julia.    If you feel this is a wanted feature I can add this to the docs.

  - **Differential Targets**: Entities defined by a time derivative (e.g., targets of SBML rate rules or species that change by participating in reactions).
  - **Algebraic Targets**: Entities defined by an algebraic assignment (i.e., they are not associated with a time derivative and are generally not constant). In the context of a neural network, if the neural network appears in the observable formula or ODE RHS, its inputs are considered algebraic targets. If a neural network is exclusively of the 3rd hybridization type (**Neural network models to parametrize ODEs**), its inputs are considered to be a constant target.
  - **Constant Targets**: Entities defined by a constant value but may be subject to event assignments (e.g., SBML model parameters that are not targets of rate or assignment rules). In the SciML PEtab standard, as input arrays files are constant, they can assign to constant targets (e.g. neural network inputs in the 3rd hybridization type).

>   - **Model Parameter Targets**: Entities corresponding to model parameters in the model file (e.g., SBML model parameters). These parameters cannot appear in the PEtab parameter table, and as discussed [here](@ref example_hybrid_output), neither in the observable formula.

Author: This depends on whether or not we keep `setParameter` (see below)

- `operationType` [STRING, REQUIRED]:
  Specifies the type of operation to be performed on the target. Repeating parts of the PEtab standard, allowed values are:
  - `setValue`: Sets the current value of the target to the value specified in `targetValue`. The target must be a constant target.
  - `setRate`: Sets the time derivative of the target to `targetValue`. The target must be a differential target.
  - `setAssignment`: Sets the target to the symbolic value of `targetValue`. The target must be an algebraic target.

>   - `setParameter`: Sets the value of a model parameter to a dynamic neural network output. `targetValue` must be a neural network output.

Author: We do care about the target. If we want to insert a neural-network into the ODE RHS (not replace a rate), we only want to change a parameter value, and we do not want to assign to a parameter given by a rate rule (because then we are setting an entire derivative). 

I think the best way to handle this is to drop `setParameter` and for replacing a parameter in the RHS use `setValue`. Then it is implied that only a constant parameter can be changed in the RHS.

>   - `setParameter`: Sets the value of a model parameter to a dynamic neural network output. `targetValue` must be a neural network output.

Reviewer: As for naming, since it changes the target type, perhaps `createAssignment` is clearer than `setParameter`. Then users might realize that the target type is changing to be time-dependent (if they're familiar with SBML...).

### The `operationType`, mapping table, and condition table defines hybrid model type

The `operationType`, mapping table, and condition table together specify where a neural network model is integrated into a PEtab SciML problem. In particular (here, we refer to neural network inputs and outputs as simply inputs and outputs):

1. **Neural network models in the ODE model’s right-hand side (RHS)**: If all inputs use `setAssignment` and all outputs use either `setRate` or `setParameter`, the neural network appears in the ODE right-hand side.
2. **Neural network models in the observable function**: If all inputs use `setAssignment` and no outputs are used in the hybridization table, then the outputs must be used in the observable formulas.
3. **Neural network models to parametrize ODEs**: If all inputs are constant targets, that is use `setValue`, are assigned in the condition table, or are assigned to a constant variable in the mapping table, and if all outputs assign using `setValue` or are assigning in the condition table, then the neural network sets model parameters or initial values before the simulation.

> All other combinations are disallowed because they generally do not make sense in a PEtab context. An important consequence of this is that if all inputs are assigned to constant targets, then neural network outputs must also be assigning to constant targets in either the hybridization or condition tables. The `petab_sciml` library provides a linter to ensure that no disallowed combination is used. For packages that do not wish to depend on Python, it is strongly recommended to verify that the input combinations in the condition table are valid.

Author: This depends on how we do with `setParameter` (see above)

> All other combinations are disallowed because they generally do not make sense in a PEtab context. An important consequence of this is that if all inputs are assigned to constant targets, then neural network outputs must also be assigning to constant targets in either the hybridization or condition tables. The `petab_sciml` library provides a linter to ensure that no disallowed combination is used. For packages that do not wish to depend on Python, it is strongly recommended to verify that the input combinations in the condition table are valid.

Reviewer: Here is a time-independent neural network that depends on an assignment rule:
- species A is determined by an assignment rule
- the neural network input the the initial condition of species A
- the initial condition of species B is the neural network output

> All other combinations are disallowed because they generally do not make sense in a PEtab context. An important consequence of this is that if all inputs are assigned to constant targets, then neural network outputs must also be assigning to constant targets in either the hybridization or condition tables. The `petab_sciml` library provides a linter to ensure that no disallowed combination is used. For packages that do not wish to depend on Python, it is strongly recommended to verify that the input combinations in the condition table are valid.

Reviewer: Here is my current understanding of the what a user can do. Rows with bold are the SciML-specific cases. Rows with italics extend operations currently supported in plain PEtab v2 conditions. I could provide a nicer version to include in the docs if useful.

Description | Target type | Supported values | Implementation
-- | -- | -- | --
Change initial condition _with neural network_ | Differential target | Any math _with a neural network output PEtab ID_ (will be evaluated at t_0) | operatorType: setValue
Change ODE _with neural network_ | Differential target | Any math _with a neural network output PEtab ID_ | operatorType: setRate
Change assignment rule _with neural network_ | Algebraic target | Any math _with a neural network output PEtab ID_ | operatorType: setAssignment
**Change observable with neural network** | PEtab observable | Any math _with a neural network output PEtab ID_ | Integrate directly into PEtab observables table.
Set parameter value _with neural network_ (time-independent) | Constant non-estimated parameter. | Any math _with a neural network output PEtab ID_ (will be evaluated at t_0) | operatorType: setValue
**Set parameter value with neural network (time-dependent)** | Constant non-estimated parameter. **Will become time-dependent!** | Any math with a neural network output PEtab ID | operatorType: setParameter

!!! note "Model structure altering assignments"

>     When `setRate`, `setAssignment`, or `setParameter` are used, the model structure or observable formula is altered during model import. Effectively, a neural network is inserted into the generated functions. PEtab SciML only supports such alterations if they are applied to all conditions; hence, these alterations can only be specified in the hybridization table.

Reviewer 2: Would drop this note. What information do we provide here that influences user behaviour?
- we no longer have operators in the condition table, so this statement will be more confusing than helpful
- if a user actually wants condition specific models, we don't really tell him how (encode multiple alternatives through parametrization and assign in condition table).

> ### [The Hybridization and the Condition table](@id hybrid_cond_tables)

Reviewer 2: why do we need this section? Are we not able to say everything we need to say in the preceding section?

Author: Yes, but with both having the hybridization table and condition table, which both can be used to assign constant targets, I think it is good to have a section that clarify how they interact.

When the neural network sets constant targets such model parameters and/or initial values prior to simulations, assignments for **all conditions** can be set with the `setValue` operator type (see example below). However, sometimes constant target assignments need to be condition-specific. In such cases, the neural network inputs and/or outputs should be assigned in the condition table rather than in the hybridization table. For example, in the output case, if `net1_output1` sets `p1` in `cond1` and `p2` in `cond2`, a valid condition table is:

> | **conditionId** | **targetId** | **targetValue** |

Reviewer 2: I thought this would be explicitly disallowed as model altering.

Author: This is not model altering, as in this scenario parameters are set prior to model simulations. So rather this is just mapping. This is already a feature that basically supported by PEtab, a parameter to be estimated can set values for several model parameters (e.g. Beer benchmark model). So I think the same should hold for neural network output (in prior to model simulation case).

|-----------------|--------------|-----------------|
| cond1           | p1           | net1_output1    |
| cond2           | p2           | net1_output1    |

Similarly, for the input case, if `net1_input` is given by `p1` in `cond1` and `p2` in `cond2`, a valid condition table is:

| **conditionId** | **targetId** | **targetValue** |
|-----------------|--------------|-----------------|
| cond1           | net1_input   | p1              |
| cond2           | net1_input   | p2              |

!!! note "Assignments for a network variables occur either in hybridization or condition table"
    Since the hybridization table sets assignments for all conditions, variables assigned in the condition table cannot be assigned in the hybridization table, and vice versa. The linter will throw an error if a variable is assigned in both tables. If all assignments occur in the condition table, the hybridization table can be left empty or omitted.

> ### [Examples: Assigning Neural Network Output](@id example_hybrid_output)

Reviewer 2: don't repeat the spec that we just introduced in the the examples. Also I would show a single self contained vignette here, rather writing this in passive voice and going through individual parts of the API. Passive is more suitable for API documentation.

Maybe it's easier to remove them for and and do another PR where we add tutorials/examples/vignettes?

Author: You mean that all example sections should be removed from the format.md file?

Author: Because, with the examples I do not want to start mixing sections as this will ultimately be confusing. So either I think examples should be kept, or we rely on tutorials for  tying everything together.

### Examples: Assigning Neural Network Input

When a neural network sets a **constant model parameter value or initial value**, its input variable (as specified in the mapping table) is treated as a standard PEtab variable. If that input variable is not defined in the parameter table, it should be assigned using the `setValue` operator type in the hybridization table if the assignment applies to all conditions; otherwise, it should be assigned in the condition table as explained [here](@ref hybrid_cond_tables).

When a neural network sets **a model derivative or alters the ODE RHS**, the input is considered to be an algebraic target. Therefore, the input variable should be assigned using the `setAssignment` operator. For example, if neural network input `net1_input1` is given by species `X`, a valid condition table is:

| **operationType** | **targetId** | **targetValue** |
|------------------|-------------|---------------|
| setAssignment | net_input1  | X             |

## [Parameter Table](@id parameter_table)

> The parameter table largely follows the same format as in PEtab version 2, with a subset of fields extended to accommodate neural network parameters. Further, a new `initializationDistribution` column is added for neural network-specific initialization. A general overview of the parameter table is available in the PEtab documentation; here, the focus is on extensions relevant to the SciML extension.

Reviewer 2: So we dropped the initialization entirely from the core spec, shouldn't we do the same with `initializationDistribution` following the argument that this is about how to train the model rather than what model to train?

Author: This is a good question. The thing is that initialization scheme matters quite a bit for neural-network training, and for reproducibility across tools it can be to test that importing tools are consistent. But yes, given the changes in the PEtab spec, it probably makes sense to remove it.

- **initializationDistribution [String, optional]**: Specifies the prior used for sampling initial values before parameter estimation. In addition to the PEtab-supported priors [ADD], the SciML extension supports the following standard neural network initialization priors:
  - [`kaimingUniform`](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.kaiming_uniform_) (default) — with `gain` as `initializationDistributionParameters` value.
  - [`kaimingNormal`](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.kaiming_normal_) — with `gain` as `initializationDistributionParameters` value.
  - [`xavierUniform`](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.xavier_uniform_) — with `gain` as `initializationDistributionParameters` value.
  - [`xavierNormal`](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.xavier_normal_) — with `gain` as `initializationDistributionParameters` value.

> - **initializationDistributionParameters [NUMERIC, optional]**: Distribution parameter value for the `initializationDistribution`. Which parameter(s) are referred to depends on the chosen prior distribution, as as in the PEtab standard, multiple parameters are separated with semicolon.

Reviewer: ```suggestion
- **initializationDistributionParameters [NUMERIC, optional]**: Distribution parameter value for the `initializationDistribution`. Which parameter(s) are referred to depends on the chosen prior distribution, as in the PEtab standard, multiple parameters are separated with semicolon.

```

### Bounds for neural net parameters

> Bounds can be specified for an entire network or its nested levels. However, it should be noted that most optimization algorithms used for neural networks, such as ADAM, do not support parameter bounds in their standard implementations. Therefore, for neural network models, bounds are optional and default to `-Inf` for the lower bound and `Inf` for the upper bound.

Reviewer: Should be lowercase `-inf` and `inf` to match PEtab (e.g. `inf` for steady-state data in the measurements table, or in v2 `-inf` for pre-equilibration)

Author: It is `Inf` in Julia, thanks for noticing I will change!

It is also possible to specify different priors for different parameter arrays within a layer. For example, to use different priors for the weights and bias in `layer1` of `net1` (assuming a `linear` layer), a valid parameter table would be:

| **parameterId**      | **parameterScale** | **lowerBound** | **upperBound** | **estimate** | **nominalValue** | **initializationDistribution** | **initializationDistributionParameters** |
|----------------------|--------------------|----------------|----------------|------------|------------------|-----------------------------|-------------------------------------|
| net1\_ps              | lin                | -inf           | inf            | 1          | netId1\_ps     | kaimingUniform              | 1                                   |
| net1\_layer1_weight   | lin                | -inf           | inf            | 1          | netId1\_ps     | kaimingUniform              | 1                                   |
| net1\_layer1_bias     | lin                | -inf           | inf            | 1          | netId1\_ps     | kaimingNormal               | 5/3                                 |

## [Problem YAML File](@id YAML_file)

> The PEtab problem YAML file follows the PEtab version 2 format, except that a mapping table is now required (it is optional in the standard). It also includes an extension section for specifying neural network YAML files, as well as any files for neural network parameters and/or inputs. Specifically the extension sections includes two new groups:

Reviewer 2: why do we wan't to explicitly say that the mapping table is required? Isn't that implied by the fact that you otherwise can't reference nn elements?

"group" is afaik non-standard YAML vocabulary, see https://yaml.org/spec/1.2.2/ext/glossary/

Author: > why do we wan't to explicitly say that the mapping table is required? Isn't that implied by the fact that you otherwise can't reference nn elements?

Fair point.

> - `extensions > petab_sciml > neural_nets`: Here each neural network is defined via a key-value mapping, where the key is the neural network Id. The corresponding value is an object whose fields depend on the network format:

Reviewer 2: Is `extensions > petab_sciml > neural_nets` a standard way of flattening YAML?

It's "mapping" and "key/value pairs". Why do we make the fields dependent on network format? can we not just always have location + language?

Author: So if the neural network now not provided via the YAML interface (e.g. Julia) should location then be empty, or should we enforce the user to code the neural network in a separate file (which is a bit unusual pattern in Julia)

Author: > Is extensions > petab_sciml > neural_nets a standard way of flattening YAML?

Fair point. I have not found how to refer, but as YAML files are dict, the following seems to be close to how it is done:

`yaml_file[extensions][petab_sciml][neural_nets]`

Reviewer 2: > So if the neural network now not provided via the YAML interface (e.g. Julia) should location then be empty, or should we enforce the user to code the neural network in a separate file (which is a bit unusual pattern in Julia)

If the model cannot be serialized, how is it supposed to be exchanged?

Author: > If the model cannot be serialized, how is it supposed to be exchanged?

Fair point, but maybe then having language is maybe a bit of a bad formulations (as there are different neural network libraries in for example Julia and Jax).  Potentially, format + location could be something (I will need to think about this)

Reviewer 2: yes, format sounds good

  - For models in the `petab_sciml` format, include the file path as `location`.
  - For neural networks in other formats, includes the neural network library name as `library`.

> - `extensions > petab_sciml > array_files`: Here each associated neural network array file. Parameter files should follow the structure [here](@ref hdf5_ps_structure), and input files the structure [here](@ref hdf5_input_structure). Each entry is specified with a key-value mapping, where the key is the array Id which is considered a valid PEtab Id. The corresponding value is an object that includes:

Reviewer 2: ```suggestion
- `extensions > petab_sciml > array_files`: Specifies array files. Parameter files should follow the structure [here](@ref hdf5_ps_structure), and input files the structure [here](@ref hdf5_input_structure). Each entry is specified with a key-value mapping, where the key is the array Id which is considered a valid PEtab Id. The corresponding value is a dict with:

```

Don't add links with body `here`, it's just fluff, directly link the thing you are referring to.

It further extends the following standard PEtab files:

1. **Mapping Table**: Used to describe how neural network inputs and outputs map to PEtab variables.
2. **Parameters Table**: Used to describe nominal values and potential priors for initializing network parameters.
3. **Problem YAML File**: Includes a new SciML field for neural network models and (optionally) array or tensor formatted data.

All other PEtab files remain unchanged. This page explains the format and options for each file that is added or modified by the PEtab SciML extension.

## High Level Overview

The aim of the PEtab SciML extension is to facilitate the creation of hybrid models that combine data-driven and mechanistic components. The extension is designed to keep the dynamic model, neural network model, and PEtab problem as independent as possible, with the models linked in the hybridization and/or condition tables.

The extension supports three hybrid model types, and a valid PEtab SciML problem can use one or any combination of these types. Each of the three types is specified differently:

> 1. **Data-driven models in the ODE model’s right-hand side (RHS):** In this scenario, the SBML file is modified during import by either replacing a derivative or setting a parameter value to a neural network output. In both cases, the neural network input and output variables (as defined in the mapping table) must be assigned in the hybridization table using the `setRate`, `setParameter` and/or `setAssignment` operator types.

Author: Thanks, agree!

!!! tip "For developers: Respect memory order"
    Tools supporting the SciML extension should, for computational efficiency, reorder input data and potential layer parameter arrays to match the memory ordering of the target language. For example, PEtab.jl converts input data to column-major order, with Julia indexing.

TODO: We will fix condition specific input in the YAML file later.

## Mapping Table

To avoid confusion regarding what a neural network ID (`netId`) refers to (e.g., parameters, inputs, etc.), `netId` is not considered a valid PEtab identifier. Consequently, every neural network input, parameter, and output must be explicitly mapped in the mapping table to a PEtab variable. In the context of the PEtab SciML extension, the relevant mapping table columns are:

- **petabEntityId [STRING]**: A valid PEtab identifier that is not defined elsewhere in the PEtab problem. This identifier can be referenced in the condition, measurement, parameter, and observable tables, but not within the model itself. For neural network outputs, the PEtab identifier must be assigned in the condition table; for inputs, this is not required (see examples below).

> - **modelEntityId [STRING]**: Describes the neural network entity corresponding to the `petabEntityId`. This must specify a parameter set (e.g. `netId.parameters`), a parameter for a specific layer (e.g. `netId.parameters.layerId`) an input (e.g. `netId.input[{n}]`), or an output (`netId.output[{n}]`), where `n` is the specific input or output index.

Author: Should we also allow neural networks to potentially have multiple outputs, and @dilpath is this possible for the yaml format?

> - **modelEntityId [STRING]**: Describes the neural network entity corresponding to the `petabEntityId`. This must specify a parameter set (e.g. `netId.parameters`), a parameter for a specific layer (e.g. `netId.parameters.layerId`) an input (e.g. `netId.input[{n}]`), or an output (`netId.output[{n}]`), where `n` is the specific input or output index.

Author: > This makes it easier for users to get an overview of rules and for developers to see how individual rules could have mutual side-effects. 

I agree, will explain all `netId.` types, and then still keep the examples as I often find that examples help clarify details.

| **petabEntityId** | **modelEntityId**   |
|-------------------|---------------------|
| net1_ps           | net1.parameters     |

Next, `net1_ps` should be assigned properties in the [parameter table](@ref parameter_table).

It is also possible to target a specific layer in the parameter table. To do so, the layer must first be mapped to a PEtab variable. For the case above, a valid mapping table to target `layer1` would be:

| **petabEntityId** | **modelEntityId**         |
|-------------------|---------------------------|
| net1\_ps           | net1.parameters          |
| net1\_ps\_layer1    | net1.parameters.layer1   |

It is also possible to target parameter arrays within a layer using the notation `netId.layerId.arrayId`.

### Additional Details

Although a neural network can, in principle, accept both array and scalar inputs for a single argument, this feature is not currently tested for among tools implementing the PEtab SciML extension due to it being hard to implement. To have both scalar and array input, the neural network model should instead have multiple arguments for its forward pass function.

## Condition and Hybridization tables

> The PEtab SciML extension introduces a new hybridization table. This table follows the same logic as the condition table but applies its assignments to all simulation conditions, so it does not include a `conditionId` column. The table is necessary because specifying which type of hybridization applies to a particular problem requires the precision of the `operatorType` column in the condition table. However, certain assignments—such as inserting a neural network into the ODE right-hand side (RHS)—apply to every simulation condition. Thus, to avoid repeating these assignments for each condition, the hybridization table is introduced.

Reviewer 2: I think we may have to change this as with the changes discussed in the last editors meeting we will no longer have operator types in the condition table and we will have to introduce the whole machinery here.

> The PEtab SciML extension introduces a new hybridization table. This table follows the same logic as the condition table but applies its assignments to all simulation conditions, so it does not include a `conditionId` column. The table is necessary because specifying which type of hybridization applies to a particular problem requires the precision of the `operatorType` column in the condition table. However, certain assignments—such as inserting a neural network into the ODE right-hand side (RHS)—apply to every simulation condition. Thus, to avoid repeating these assignments for each condition, the hybridization table is introduced.

Author: I know, I have changed and this will be in the next version.

All other PEtab files remain unchanged. This page explains the format and options for each file that is added or modified by the PEtab SciML extension.

## High Level Overview

The aim of the PEtab SciML extension is to facilitate the creation of hybrid models that combine neural network and mechanistic components. The extension is designed to keep the dynamic model, neural network model, and PEtab problem as independent as possible, with the models linked in the hybridization and/or condition tables.

The extension supports three hybrid model types, and a valid PEtab SciML problem can use one or any combination of these types. The three types are:

1. **Neural network models in the ODE model’s right-hand side (RHS):** In this scenario, the model (e.g. SBML) file is modified during import by either replacing a derivative or setting a parameter value to a neural network output.
2. **Neural network models in the observable function:** In this scenario, the neural network output variable (as defined in the mapping table) is directly inserted in the observable formula.

> 3. **Neural network models to parametrize ODEs:** In this scenario, the neural network model sets constant parameters or initial values in the ODE model prior to simulation.

Author: I think case 4 should be covered by 1-3. As in 1-2 we have the the ODE (or rather model) sets the input, and in 3 we have that potentially the inputs are set by PEtab variables. So I would not add 4, as I think the three cases we have are observable function, before ODE, and inside ODE, which is 1-3.

> 3. **Neural network models to parametrize ODEs:** In this scenario, the neural network model sets constant parameters or initial values in the ODE model prior to simulation.

Reviewer: I'm not sure. If I ask "Can neural network inputs be data values or PEtab variables?", I don't think I see the answer in cases 1-3.

> 3. **Neural network models to parametrize ODEs:** In this scenario, the neural network model sets constant parameters or initial values in the ODE model prior to simulation.

Author: Which is fair, but I do not see this as a class of hybrid model, rather, I see it as one of the ways to configure approach 3 (and I believe that with the rest of the docs it should be clear how inputs can be treated)

## [Neural Network Model Format](@id net_format)

The neural network model format is flexible, meaning that models can be provided in any format supported by tools compatible with the PEtab SciML format (for example, [Lux.jl](https://github.com/LuxDL/Lux.jl) in [PEtab.jl](https://github.com/sebapersson/PEtab.jl)). Additionally, the `petab_sciml` library offers a neural network YAML file format, which can be imported into tools across programming languages. This format flexibility exists because, although the YAML format can accommodate many architectures, some may still be difficult to represent. However, when possible, we recommend using the YAML format to facilitate model exchange across different software.

Regardless of the model format, to be compatible with the other files in a PEtab SciML problem a neural network model must consist of two parts:

> - **layers**: A constructor that defines the network layers, each with a unique identifier.

Author: I would say no. I think we can keep input as an optional argument, as this can allow for nice referencing in the mapping table (your comment below).

> - **forward**: A forward pass function that, given input arguments, specifies the order in which layers are called, applies any activation functions, and returns a single array as output. The forward function can accept more than one input argument (`n > 1`), and in the [mapping table](@ref mapping_table), the forward function's `n`th input argument (ignoring any potential class arguments such as `self`) is considered as argument `n`.

Author: I do not think I follow what you mean here?

> - **forward**: A forward pass function that, given input arguments, specifies the order in which layers are called, applies any activation functions, and returns a single array as output. The forward function can accept more than one input argument (`n > 1`), and in the [mapping table](@ref mapping_table), the forward function's `n`th input argument (ignoring any potential class arguments such as `self`) is considered as argument `n`.

Reviewer: It's technically possible to define layers in pytorch's `forward`, but then I think they're not trainable. It's similar for us (with our neural network YAML format), so I tried to write that explicitly here.

> - **forward**: A forward pass function that, given input arguments, specifies the order in which layers are called, applies any activation functions, and returns a single array as output. The forward function can accept more than one input argument (`n > 1`), and in the [mapping table](@ref mapping_table), the forward function's `n`th input argument (ignoring any potential class arguments such as `self`) is considered as argument `n`.

Author: Aha, basically that we do not allow non trainable (frozen) layers?

> - **forward**: A forward pass function that, given input arguments, specifies the order in which layers are called, applies any activation functions, and returns a single array as output. The forward function can accept more than one input argument (`n > 1`), and in the [mapping table](@ref mapping_table), the forward function's `n`th input argument (ignoring any potential class arguments such as `self`) is considered as argument `n`.

Reviewer: I think it's OK to allow frozen layers. Just my point is:
- is it clear to users that things declared in `forward` cannot be trained? This is why I wrote that things with estimated parameters should be declared in `layers` first, before using them in `forward`.
- is it clear to users that things in `forward` currently cannot be referenced in other PEtab files? This is why I wrote that they are not visible.

> - **forward**: A forward pass function that, given input arguments, specifies the order in which layers are called, applies any activation functions, and returns a single array as output. The forward function can accept more than one input argument (`n > 1`), and in the [mapping table](@ref mapping_table), the forward function's `n`th input argument (ignoring any potential class arguments such as `self`) is considered as argument `n`.

Author: Ahh, ok I think these are fair points, thanks!

└───layer1 (group)
│   ├── arrayId{1}
│   └── arrayId{2}
└───layer2 (group)
    ├── arrayId{1}
    └── arrayId{2}
```

Here, `arrayId` depends on the naming convention for parameters in each layer. For example, a PyTorch `linear` layer typically has arrays named `weight` and (optionally) `bias`. While these names are common in many layers, the actual `arrayId` depends on the layer type and the specific neural network library.

Because parameters are stored in HDF5, they are saved as arrays. The indexing convention and naming therefore depends on the model library:

- For neural network models in the PEtab SciML YAML format, indexing follows PyTorch conventions. Usually, users do not need to handle these details directly, as PEtab SciML tools manage them automatically. This also means that `arrayId` follows PyTorch naming convention.
- For neural networks provided by another library, the indexing and `arrayId` naming follow that library’s conventions.

### [Neural Network Input](@id hdf5_input_structure)

When network input is provided as an array via an HDF5 file (see the mapping table below), its format should be:

```hdf5
input.hdf5
└───input (group)
    └─── input_array

> ```

Author: Thanks for the heads up. As we also support neural networks from other packages, where inputs do not have ids, I think in the end we need to aim for a solution where inputs can have ids, or indices as current described in the docs.

    Tools supporting the SciML extension should, for computational efficiency, reorder input data and potential layer parameter arrays to match the memory ordering of the target language. For example, PEtab.jl converts input data to column-major order, with Julia indexing.

TODO: We will fix condition specific input in the YAML file later.

## [Mapping Table](@id mapping_table)

To avoid confusion regarding what a neural network Id (`netId`) refers to (e.g., parameters, inputs, etc.), `netId` is not considered a valid PEtab identifier. Consequently, every neural network input, parameter, and output must be explicitly mapped in the mapping table to a PEtab variable.

### Detailed Field Description

- **petabEntityId** [STRING, required]: A valid PEtab identifier that is not defined elsewhere in the PEtab problem. This identifier can be referenced in the condition, hybridization, measurement, parameter, yaml file (in case of file input), and observable tables, but not within the model itself.
- **modelEntityId** [STRING, required]: Describes the neural network component mapped to the `petabEntityId`. For a neural network, the valid identifiers are:  

>   - `netId.parameters`: Parameters for a neural network model. A specific layer can also be referenced with `netId.parameters.layerId`, and specific arrays in a layer can be referenced with `netId.parameters.arrayId`. For parameter arrays, individual indexing is not allowed.

Author: This is a a good point. I also like the first approach to referencing, and would like to change to that. @FFroehlich what do you think?

>   - `netId.parameters`: Parameters for a neural network model. A specific layer can also be referenced with `netId.parameters.layerId`, and specific arrays in a layer can be referenced with `netId.parameters.arrayId`. For parameter arrays, individual indexing is not allowed.

Reviewer 2: Agreed, is the `.parameters` suffix necessary though?

>   - `netId.parameters`: Parameters for a neural network model. A specific layer can also be referenced with `netId.parameters.layerId`, and specific arrays in a layer can be referenced with `netId.parameters.arrayId`. For parameter arrays, individual indexing is not allowed.

Reviewer 2: right, but `netId.layers[layerID].weights` and `netId.layers[layerID].bias` should be fine, right?

>   - `netId.parameters`: Parameters for a neural network model. A specific layer can also be referenced with `netId.parameters.layerId`, and specific arrays in a layer can be referenced with `netId.parameters.arrayId`. For parameter arrays, individual indexing is not allowed.

Author: The latter should be fine, but maybe we should add the parameters ending just to be explicit?

>   - `netId.parameters`: Parameters for a neural network model. A specific layer can also be referenced with `netId.parameters.layerId`, and specific arrays in a layer can be referenced with `netId.parameters.arrayId`. For parameter arrays, individual indexing is not allowed.

Author: > netId.layers[layerId].weights, netId.layers[layerId].biases, and netId.layers[layerId].parameters is best. 

I thing this sounds good, and it will allow us to keep flexibility later with potentially strange outputs.

  - `netId.input`: Input for a neural network with a single input argument (as in the first example [here](@ref YAML_net_format)). If the input is provided as an array from a file, use `netId.input`. Otherwise, the input is assumed to be a `Vector`, where each element `n` should be mapped to a PEtab ID, with element `n` specified as `netId.input[{n}]`.

>   - `netId.inputs`: Inputs for a neural network with multiple input arguments (as in the second example [here](@ref YAML_net_format)). Each input argument is accessed via `netId.inputs[{n}]`, and for each argument the same rules apply as for `netId.input`. For example, to access element `m` in input `n`, write `netId.inputs[{n}][{m}]`.

Author: For inputs, I agree we could use something like:
`netId.inputs.Id...`

or

`netId.inputs[{n}]`

and similarly for outputs. However, for layers (parameters), I prefer not to enforce indexing. Order is crucial for inputs and outputs to correctly provide the arguments for the forward trace function, but for layers, enforcing a specific order would only complicate implementation, especially since most packages store neural network parameters in a dictionary-like structure.

Before implementing this change though, what do you think @FFroehlich

>   - `netId.inputs`: Inputs for a neural network with multiple input arguments (as in the second example [here](@ref YAML_net_format)). Each input argument is accessed via `netId.inputs[{n}]`, and for each argument the same rules apply as for `netId.input`. For example, to access element `m` in input `n`, write `netId.inputs[{n}][{m}]`.

Reviewer 2: same as above, why do we need `.values` and `.parameters`?

>   - `netId.inputs`: Inputs for a neural network with multiple input arguments (as in the second example [here](@ref YAML_net_format)). Each input argument is accessed via `netId.inputs[{n}]`, and for each argument the same rules apply as for `netId.input`. For example, to access element `m` in input `n`, write `netId.inputs[{n}][{m}]`.

Author: I think values in this case is redundant, so would remove it.

>   - `netId.inputs`: Inputs for a neural network with multiple input arguments (as in the second example [here](@ref YAML_net_format)). Each input argument is accessed via `netId.inputs[{n}]`, and for each argument the same rules apply as for `netId.input`. For example, to access element `m` in input `n`, write `netId.inputs[{n}][{m}]`.

Reviewer: Yes values is a bit redundant. I don't like the multi-indexing like `netId.inputs[input index][value index]` because it looks a bit like an array when it isn't. Also `.values` seemed more consistent with treating `net`/`layer`/`input`/`output` as objects with attributes (e.g., `netId.parameters`, not `netId`, to get the parameters), but no strong opinion, fine to remove `.values`.

As soon as there might be multiple properties for `inputs` and `outputs`, then `.values` makes sense just like for `netId` vs. `netId.parameters`. e.g. `input.values` and `input.source`, where source is a HDF5 data file, is conceivable to me. But again, just a minor thing that we can revisit later if it's an issue

  - `netId.output`: Neural network output, assumed to be a `Vector`, where each element `n` should be mapped to a PEtab ID, with element `n` given by `netId.input[{n}]`.

### Example: Network with Scalar Inputs

For networks with scalar inputs, the PEtab entity should be a PEtab variable. For example, assume that the network `net1` has two inputs, then a valid mapping table would be:

| **petabEntityId** | **modelEntityId** |
|-------------------|-------------------|
| net1\_input1       | net1.input[1]       |
| net1\_input2       | net1.input[2]       |

Scalar input variables can then be:

> - **Parameters in the parameters table**: These may be either estimated or constant.

Author: It is a bit strange, but in general a valid PEtab operation. I would still keep it, have test-cases on it, and then let software be free to skip it if the feature is to tricky to implement.

> - **Parameters assigned in the condition or hybridization table**: More details on this option can be found in the section on the condition and hybridization tables.

Author: The condition table is not irrelevant, as for inputs we can still have condition specific inputs for neural networks before the ODE model (in this case the input is a constant target). I will double check this part of the spec, and see if I can make it more clear.

> - **Parameters assigned in the condition or hybridization table**: More details on this option can be found in the section on the condition and hybridization tables.

Reviewer: Yes you're right, not sure what that point was anymore...

Could change this to link to the section: "More details on this option can be found in the section on the condition and hybridization tables."

P.S. I don't see "the section on the condition and hybridization tables" in the table of contents in the left panel.
https://html-preview.github.io/?url=https://raw.githubusercontent.com/sebapersson/petab_sciml/c5251c10fb7577d239eb643ea1e7b066f6d76bd7/docs/build/format.html#hybrid_table

TODO: check that it appears in the published version.

> - **Parameters assigned in the condition or hybridization table**: More details on this option can be found in the section on the condition and hybridization tables.

Author: It is a sub-section, but the link is correct :)

| **petabEntityId** | **modelEntityId** |
|-------------------|-------------------|
| net1\_input        | net1.input        |

This variable (here `net1_input`) should then assigned to specific input file variables (e.g. `net1_input_cond1` and `net1_input_cond2`) via the condition table. For a full example of a valid PEtab problem with array inputs, see [ADD].

TODO: Will fix an input format later.

### [Example: Network Observable Formula Output](@id output_obs)

If the neural network output appears in the observable formula, the PEtab entity should be directly referenced in the observable formula. For example, given:

| **petabEntityId** | **modelEntityId**  |
|-----------------|-----------------|
| net1\_output1    | net1.output[1]    |

A valid entry in the observable table would be:

| **observableId** | **observableFormula** |
|----------------|----------------------|
| obs1           | net1\_output1        |

As usual, the `observableFormula` can be any valid PEtab equation, so `net1_output1 + 1` would also be valid.

> ### Example: Network Scalar Output

Author: I think I will wait to after I have moved the docs to Python to see which formatting options works the best.

  - **Model Parameter Targets**: Entities corresponding to model parameters in the model file (e.g., SBML model parameters).
- `operationType` [STRING, required]:
  Specifies the type of operation to be performed on the target. Allowed values are:
  - `setValue`: Sets the current value of the target to the value specified in `targetValue`. The target must be a constant target.
  - `setRate`: Sets the time derivative of the target to `targetValue`. The target must be a differential target.
  - `setAssignment`: Sets the target to the symbolic value of `targetValue`. The target must be an algebraic target.
  - `setParameter`: Assigns the value of a model parameter to a neural network output. `targetValue` must be a neural network output.
- `targetValue` [STRING, required]:
  The value or expression that will be used to change the target. The interpretation of this value depends on the specified `operationType`.

### `operationType` and the Condition table Defines Hybrid Model Type

The `operationType` and condition table together specify where a neural network model is integrated into a PEtab SciML problem. In particular:

> - **ODE RHS**: If all inputs use `setAssignment` and all outputs use either `setRate` or `setParameter`, the neural network appears in the ODE right-hand side.

Author: This is a good point, thanks!

### Examples: Assigning Neural Network Output

To set a **constant model parameter value** before model simulations for all conditions, the `setValue` operator type should be used. For example, if parameter `p` is determined by `net1_output1` (mapped to a neural network output in the mapping table) for all conditions, a valid hybridization table entry is:

| **operationType** | **targetId** | **targetValue** |
|------------------|--------------|-----------------|
| setValue         | p            | net1_output1    |

Similar, to set an **initial value**, the `setValue` operator type should be used. For example, if the initial value of species `X` is determined by `net1_output1` for all conditions, a valid hybridization table entry is:

| **operationType** | **targetId** | **targetValue** |
|------------------|--------------|-----------------|

> | setInitial         | X            | net1_output1    |

Author: I like `setInitial`, but I forgot to remove it here. As `setValue` only applies to constant targets, if a specie is then given as the target, it can only be the initial value. As I think we should avoid introducing another `operatorType`, I will simply add `setValue` here.

These assignments above hold for all conditions. As explained [here](@ref hybrid_cond_tables), condition-specific assignments are also possible when the neural network sets variables prior to model simulations.

To set a **model derivative**, the `setRate` operator type should be used. For example, if the rate of species `X` is determined by `net1_output1`, a valid hybridization table entry is:

| **operationType** | **targetId** | **targetValue** |
|------------------|--------------|-----------------|
| setRate          | X            | net1_output1    |

To alter the **ODE RHS**, the `setParameter` operator type should be used.  For example, if an ODE model parameter `p` should be given by `net1_output1`, a valid hybridization table entry is:

| **operationType**    | **targetId** | **targetValue** |
|---------------------|--------------|-----------------|
| setParameter    | p            | net1\_output1    |

> In theory for this last example, the `targetId` (here `p`) can also appear in the observable formula. However, as described [here](@ref output_obs), for the observable formula the mapped output variable must be directly encoded in the formula. Therefore, `targetId` appearing in the observable formula is not allowed, and the `petab_sciml` linter checks for this.

Author: This is supported, since p is static from the point of assignments (so its value us correct in the observable formula). I think the best way to formulate here is probably to say that if p is a model parameter target, it cannot appear in the observable formula.

### Examples: Assigning Neural Network Input

When a neural network sets a **constant model parameter value or initial value**, its input variable (as specified in the mapping table) is treated as a standard PEtab variable. If that input variable is not defined in the parameter table, it should be assigned using the `setValue` operator type in the hybridization table if the assignment applies to all conditions; otherwise, it should be assigned in the condition table as explained [here](@ref hybrid_cond_tables).

When a neural network sets **a model derivative or alters the ODE RHS**, the input is considered to be an algebraic target. Therefore, the input variable should be assigned using the `setAssignment` operator. For example, if neural network input `net1_input1` is given by specie `X`, a valid condition table is:

| **operationType** | **targetId** | **targetValue** |
|------------------|-------------|---------------|
| setAssignment | net_input1  | X             |

## [Parameter Table](@id parameter_table)

> The parameter table largely follows the same format as in PEtab version 2, with a subset of fields extended to accommodate neural network parameters. Further, a new `initializationPriorType` column is added for neural network-specific initialization. A general overview of the parameter table is available in the PEtab documentation; here, the focus is on extensions relevant to the SciML extension.

Author: But it is a distribution, how about `initializationDistribution`?

> The parameter table largely follows the same format as in PEtab version 2, with a subset of fields extended to accommodate neural network parameters. Further, a new `initializationPriorType` column is added for neural network-specific initialization. A general overview of the parameter table is available in the PEtab documentation; here, the focus is on extensions relevant to the SciML extension.

Reviewer: Also fine for me, since there is no plan to support things like latin hypercube / sobol sequences here :+1:

!!! note "Specific Assignments Have Precedence"
    When parsing, more specific assignments (e.g., `netId.layerId` rather than `netId`) take precedence for nominal values, priors, and so on. This means that if `netId` has one prior, and `netId.layerId` has another, the more specific assignment `netId.layerId` overrides the less specific one `netId` for `layerId` in this case.

### Detailed Field Description

- **parameterId [String, required]**: Identifies the neural network or a specific layer/parameter array. The target of the `parameterId` must be assigned via the [mapping table](@ref mapping_table).
- **nominalValue [String | NUMERIC, required]**: Specifies neural network nominal values. This can be:

>   - A PEtab variable that via the problem [YAML file](@ref YAML_file) maps to an HDF5 file with the required [structure](@ref hdf5_ps_structure). If no file exists at the given path when the problem is imported and the parameters are set to be estimated, a file is created with randomly sampled values. Unless a numeric value is provided, referring to the same file is required for all assignments for a neural network, since all neural network parameters should be collected in a single HDF5 file following the structure described [here](@ref hdf5_ps_structure).

Author: Input structure should already be specified. A bit unsure what you mean.

>   - A PEtab variable that via the problem [YAML file](@ref YAML_file) maps to an HDF5 file with the required [structure](@ref hdf5_ps_structure). If no file exists at the given path when the problem is imported and the parameters are set to be estimated, a file is created with randomly sampled values. Unless a numeric value is provided, referring to the same file is required for all assignments for a neural network, since all neural network parameters should be collected in a single HDF5 file following the structure described [here](@ref hdf5_ps_structure).

Reviewer: I meant, if neural network inputs are data from a file, where is that data file linked to the inputs in the PEtab problem?

>   - A PEtab variable that via the problem [YAML file](@ref YAML_file) maps to an HDF5 file with the required [structure](@ref hdf5_ps_structure). If no file exists at the given path when the problem is imported and the parameters are set to be estimated, a file is created with randomly sampled values. Unless a numeric value is provided, referring to the same file is required for all assignments for a neural network, since all neural network parameters should be collected in a single HDF5 file following the structure described [here](@ref hdf5_ps_structure).

Author: They are linked either in the hybridization or condition table, I clarified this now under the mapping table section.

>   - A PEtab variable that via the problem [YAML file](@ref YAML_file) maps to an HDF5 file with the required [structure](@ref hdf5_ps_structure). If no file exists at the given path when the problem is imported and the parameters are set to be estimated, a file is created with randomly sampled values. Unless a numeric value is provided, referring to the same file is required for all assignments for a neural network, since all neural network parameters should be collected in a single HDF5 file following the structure described [here](@ref hdf5_ps_structure).

Author: How about I write something like this as an addition in the mapping table section:

Sometimes, such as with image data, a neural network requires array input. In these situations, the entire input (`netId.input`) is mapped to a PEtab variable, and the PEtab variable is then assigned to a file in either the condition or hybridization table. For example, given that `net1` has a file input, a valid mapping table would be:

| **petabEntityId** | **modelEntityId** |
|-------------------|-------------------|
| net1_input        | net1.input        |

This variable (here `net1_input`) should then be mapped to specific input file variables defined in the [yaml](@ref YAML_file) problem file via the condition or hybridization table. Note that this approach allows for both condition-specific and condition-constant input files.

>   - A PEtab variable that via the problem [YAML file](@ref YAML_file) maps to an HDF5 file with the required [structure](@ref hdf5_ps_structure). If no file exists at the given path when the problem is imported and the parameters are set to be estimated, a file is created with randomly sampled values. Unless a numeric value is provided, referring to the same file is required for all assignments for a neural network, since all neural network parameters should be collected in a single HDF5 file following the structure described [here](@ref hdf5_ps_structure).

Reviewer: Ah it's clear in it's current state, thanks

- **initializationPriorType [String, optional]**: Specifies the prior used for sampling initial values before parameter estimation. In addition to the PEtab-supported priors [ADD], the SciML extension supports the following standard neural network initialization priors:
  - [`kaimingUniform`](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.kaiming_uniform_) (default) — with `gain` as `initializationPriorParameters` value.
  - [`kaimingNormal`](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.kaiming_normal_) — with `gain` as `initializationPriorParameters` value.
  - [`xavierUniform`](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.xavier_uniform_) — with `gain` as `initializationPriorParameters` value.
  - [`xavierNormal`](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.xavier_normal_) — with `gain` as `initializationPriorParameters` value.
- - **initializationPriorParameters [NUMERIC, optional]**: Distribution parameter value for the `initializationPriorType`. Which parameter(s) are referred to depends on the chosen prior distribution.

### Bounds for neural net parameters

> Bounds can be specified for an entire network or its nested levels. However, it should be noted that most optimization algorithms used for neural networks, such as ADAM, do not support parameter bounds in their standard implementation.

Author: Good question. I think the best solution is to make them optional, and let them default to `-Inf` and `Inf`.

As for PEtab v2, I think we should keep bounds, as this is an important reproducibility setting.

| net1\_layer2_ps     | lin                | -inf           | inf            | 1          | net1\_ps\_file      | kaimingUniform              | 5/3                                 |

Where `parameterId` are assumed to have been properly assigned in the [mapping table](@ref mapping_table). In this example, each layer references the same file variable for `nominalValue`. This means the layers obtain their values from the specified file.

It is also possible to specify different priors for different parameter arrays within a layer. For example, to use different priors for the weights and bias in `layer1` of `net1` (assuming a `linear` layer), a valid parameter table would be:

| **parameterId**      | **parameterScale** | **lowerBound** | **upperBound** | **estimate** | **nominalValue** | **initializationPriorType** | **initializationPriorParameters** |
|----------------------|--------------------|----------------|----------------|------------|------------------|-----------------------------|-------------------------------------|

> | net1\_ps              | lin                | -inf           | inf            | 1          | net1\_ps\_file      | kaimingUniform              | 1                                   |

Author: Good point, thanks!

```yaml
extensions:
  petab_sciml:
    netId1:
      location: file_path1.yaml
    netId2:
      location: file_path2.yaml
    array_files:
      netId1_input:
        location: netId1_input.hdf5
        language: hdf5
      netId1_ps:
        location: netId1_ps.hdf5
        language: hdf5
```

Here, `netId1` and `netId2` are the IDs of the neural network models, and `net1_input` and `net1_ps` corresponding to array files that can be used in the PEtab tables. In this case, `net1_input` would be used in the mapping table for `netId1`’s input, while `netId1_ps` would be the `nominalValue` in the parameter table.

If the neural network is provided in another format—typically one specific to a certain implementation—the neural network library should be provided to inform users which library is used. For example, when using Lux.jl in Julia, a valid file would be:

```yaml
extensions:
  petab_sciml:
    netId1:

>       library: Lux.jl

Author: Ahh fair point, both PEtab.jl and Lux.jl are required. I will clarify

>       library: Lux.jl

Reviewer: I mean, we have a clear understanding of how our neural network YAML file integrates into the PEtab problem. Is it clear how a plain Lux.jl model would integrate into the PEtab problem? If it is not clear, then PEtab.jl and PEtab_competitor.jl might make some opinionated implementations that contradict each other.

I can't think of a good example, so here is a hypothetical. If layer IDs are not supported by Lux.jl, then PEtab.jl might introduce that in it's Lux.jl importer. But then, PEtab_competitor.jl will need to not only implement a Lux.jl importer, but also implement the same layer ID format as PEtab.jl.

Not sure if I'm clear, but if you disagree that it's an issue, or the issue is unclear, then fine for me to leave this for now. I doubt it will become a real issue for us in practice.

>       library: Lux.jl

Author: Lux.jl has layer-ids, so I think it should be clear. I think I write somewhere in the spec as well that even custom formats must support ids.

    netId2:
      library: Lux.jl
```

It is then up to the specific implementation to provide the neural network model during import of the PEtab problem.

> Any number of neural networks can be specified. For example, for a model with a single neural network with ID `net1`, a valid YAML file would be:

Author: Never explicitly. I will add a short section on this.

> Any number of neural networks can be specified. For example, for a model with a single neural network with ID `net1`, a valid YAML file would be:

Author: How about adding something like this:

As mentioned, one or any combination of the hybridization types is allowed. Aside from ensuring that neural networks do not conflict (for example, by having the same target), no special considerations are required to insert multiple networks into a problem. Each additional network is added just as in the single-network case. The documentation below explains how to insert a single network.

> Any number of neural networks can be specified. For example, for a model with a single neural network with ID `net1`, a valid YAML file would be:

Reviewer: Yes, works for me :+1:

### [Neural Network Input](@id hdf5_input_structure)

When network input is provided as an array via an HDF5 file (see the mapping table below), its format should be:

```hdf5
input.hdf5
└───input (group)
    └─── input_array
```

As with [parameters](@ref hdf5_ps_structure), the indexing depends on the neural network library:

- For neural network models in the PEtab SciML format, indexing follows PyTorch conventions. For example, if the first layer is `Conv2d`, the input should be in `(C, W, H)` format, with data stored in row-major order. In general, the input should be structured to be directly compatible with PyTorch.
- For neural networks provided by another library, the indexing and ordering follow the conventions of that library.

!!! tip "For developers: Respect memory order"
    Tools supporting the SciML extension should, for computational efficiency, reorder input data and potential layer parameter arrays to match the memory ordering of the target language. For example, PEtab.jl converts input data to column-major order, with Julia indexing.

TODO: We will fix condition specific input in the YAML file later.

## [Mapping Table](@id mapping_table)

> Each neural network is assigned an Id in the PEtab SciML extension section of the PEtab problem [YAML](@ref YAML_file) file. To avoid confusion about what a neural network Id refers to (e.g., parameters, inputs, outputs), a neural network Id is not a valid PEtab identifier. Therefore, every neural network input, parameter, and output that will be referenced in the PEtab problem must be mapped to PEtab identifiers problem via the mapping table. Repeating parts of the PEtab standard, the key columns in the mapping table are:

Reviewer: ```suggestion
Each neural network is assigned an Id in the PEtab SciML extension section of the PEtab problem [YAML](@ref YAML_file) file. To avoid confusion about what a neural network Id refers to (e.g., parameters, inputs, outputs), a neural network Id is not a valid PEtab identifier. Therefore, every neural network input, parameter, and output that will be referenced in the PEtab problem must be mapped to PEtab identifiers via the mapping table. Repeating parts of the PEtab standard, the key columns in the mapping table are:

```

# Format Specification

A PEtab SciML problem extends the PEtab standard version 2 to accommodate hybrid models (SciML problems) that combine neural network and mechanistic components. Two new file types are introduced by the extension:

1. [Neural Net File(s)](@ref net_format): Optional YAML file(s) describing neural net model(s).

> 2. [Hybridization table](@ref hybrid_table): Table for assigning neural network outputs and inputs.

Author: Lets add this in the PR where we add the input HDF5 format.

2. [Parameters Table](@ref parameter_table): Extended to describe nominal values for network parameters.
3. [Problem YAML File](@ref YAML_file): Extended to include a new SciML field for neural network models and (optionally) array or tensor formatted data.

All other PEtab files remain unchanged. This specification explains the format for each file that is added or modified by the PEtab SciML extension.

## [High Level Overview](@id hybrid_types)

The PEtab SciML specification is designed to keep the dynamic model, neural network model, and PEtab problem as independent as possible while linking them through the hybridization and/or condition tables. In this context, mechanistic models are typically defined using community standards like SBML and are commonly simulated as systems of ordinary differential equations (ODEs), and here the terms mechanistic model and ODE are used interchangeably. Essentially, the PEtab SciML approach takes a PEtab problem involving a mechanistic ODE model and supports the integration of neural network inputs and outputs.

PEtab SciML supports two classes of hybrid models:

> 1. **Pre simulation hybridization**: The neural network model sets constant parameters and/or initial values in the ODE model prior to model simulation. Inputs are constant per simulation condition.

Author: Thanks, I will reword.

| **targetId**                | **targetValue**     |
|-------------------------|---------------|
| NON\_ESTIMATED\_ENTITY\_ID | MATH\_EXPRESSION |
| net1\_input1             | p1              |
| net1\_input2             | p1              |
| ...                     | ...             |

### Detailed Field Description

- `targetId` [NON\_ESTIMATED\_ENTITY\_ID, REQUIRED]: The identifier of the non-estimated entity that will be modified. Restrictions depend on hybridization type ([pre- or intra-simulation hybridization](@ref hybrid_types)). See below.
- `targetValue` [STRING, REQUIRED]: The value or expression that will be used to change the target.

### Pre-simulation hybridization

Pre-simulation neural network model inputs and outputs are constant targets (see case 1 [here](@ref hybrid_types)).

#### Inputs

Valid `targetValue`'s for a neural network input are:

- A parameter in the parameter table.
- An array input file (assigned an Id in the [YAML problem file](@ref YAML_file)).

#### Outputs

Valid `targetId`'s for a neural network output are:

- A non-estimated model parameter.

> - A specie's initial value (referenced by the specie's Id). In this case, any other specie initialization is overridden.

Reviewer: ```suggestion
- A species' initial value (referenced by the species' Id). In this case, any other species initialization is overridden.
```

- Refer to another PEtab identifier with the same nested structure, or
- Follow the corresponding hierarchical HDF5 [input](@ref hdf5_input_structure) or [parameter](@ref hdf5_ps_structure) structure.

## [Hybridization Table](@id hybrid_table)

A tab-separated values file for assigning neural network inputs and outputs. Assignments in the table the table apply to all simulation conditions. Expected to have, in any order, the following two columns:

| **targetId**                | **targetValue**     |
|-------------------------|---------------|
| NON\_ESTIMATED\_ENTITY\_ID | MATH\_EXPRESSION |
| net1\_input1             | p1              |
| net1\_input2             | p1              |
| ...                     | ...             |

### Detailed Field Description

- `targetId` [NON\_ESTIMATED\_ENTITY\_ID, REQUIRED]: The identifier of the non-estimated entity that will be modified. Restrictions depend on hybridization type ([pre- or intra-simulation hybridization](@ref hybrid_types)). See below.
- `targetValue` [STRING, REQUIRED]: The value or expression that will be used to change the target.

### Pre-simulation hybridization

> Pre-simulation neural network model inputs and outputs are constant targets (see case 1 [here](@ref hybrid_types)).

Author: I was initially a bit hesitant about this as it makes it sound like the neural network is dynamic (which only a subset of architectures, e.g. LSTM, are). But with the spec we are talking about hybridization, so dynamic hybridization does make sense.

@dilpath what do you think, static and dynamic?

> Pre-simulation neural network model inputs and outputs are constant targets (see case 1 [here](@ref hybrid_types)).

Reviewer: I think "initial" and "dynamic" hybridization types make the most sense to me, since this aligns well with ODE terminology.

Since the types are mutually exclusive (a net cannot be used for both initial and dynamic hybridization, right?), if dynamic is problematic then "initial" and "noninitial" would also be fine for me, since the hybridization type is currently a binary variable.

> Pre-simulation neural network model inputs and outputs are constant targets (see case 1 [here](@ref hybrid_types)).

Reviewer 2: "initial hybridization" and "initial neural network" don't sounds quite right to me as it suggests that the neural network/hybridization is dynamic, which is the point that sebastian raised. It would have to be "initialising" and then it still sounds like it's only applied at the beginning of an experiment. 

I would still prefer static/dynamic.

> Pre-simulation neural network model inputs and outputs are constant targets (see case 1 [here](@ref hybrid_types)).

Author: Hmm, I say we go for static/dynamic for the moment. If we come up with something we like better in a future PR it is an easy fix

> Pre-simulation neural network model inputs and outputs are constant targets (see case 1 [here](@ref hybrid_types)).

Reviewer: Also fine for me :+1:
